{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculating various statistics to investigate admixture ratios."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run ../scripts/notebook_settings.py\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['savefig.dpi'] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data_samples = pd.read_table(\"../data/metadata_with_x_missing.txt\", sep=\" \")\n",
    "chrom_number = \"X\"\n",
    "zarr_dir = \"/faststorage/project/baboondiversity/data/PG_baboons_mmul_20_04_2021/callset.zarr/chr{}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, running D on chr2 with many possible hypothesis, to make sure that the results are the same as in fs03"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "callset = zarr.open_group(zarr_dir.format(2), mode=\"r\")\n",
    "gt = allel.GenotypeArray(callset[\"calldata/GT\"])\n",
    "g_ac = gt.take(meta_data_samples.loc[meta_data_samples.C_origin == \"Gelada, Captive\"].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "def d_run(a, b, c):\n",
    "    a_ids = meta_data_samples.loc[meta_data_samples.C_origin == a].callset_index.values\n",
    "    b_ids = meta_data_samples.loc[meta_data_samples.C_origin == b].callset_index.values\n",
    "    c_ids = meta_data_samples.loc[meta_data_samples.C_origin == c].callset_index.values\n",
    "    a_ac = gt.take(a_ids, axis=1).count_alleles(max_allele=1)\n",
    "    b_ac = gt.take(b_ids, axis=1).count_alleles(max_allele=1)\n",
    "    c_ac = gt.take(c_ids, axis=1).count_alleles(max_allele=1)\n",
    "    d_stats = allel.average_patterson_d(a_ac, b_ac, c_ac, g_ac, blen=100000)\n",
    "    return d_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "origin_list = [\"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Hamadryas, Ethiopia\",\n",
    "               \"Cynocephalus, Central Tanzania\", \"Cynocephalus, Western Tanzania\",\n",
    "               \"Ursinus, Zambia\", \"Kindae, Zambia\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/eriks/miniconda3/envs/baboondiversity/lib/python3.7/site-packages/allel/stats/admixture.py:434: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  z = d_avg / se\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "a_l = []\n",
    "b_l = []\n",
    "c_l = []\n",
    "dbranch_l = []\n",
    "z_l = []\n",
    "\n",
    "for i in range(len(origin_list)):\n",
    "    for j in range(i):\n",
    "        for x in origin_list:\n",
    "            d_stats = d_run(x, origin_list[i], origin_list[j])\n",
    "            a_l.append(x)\n",
    "            b_l.append(origin_list[i])\n",
    "            c_l.append(origin_list[j])\n",
    "            dbranch_l.append(d_stats[0])\n",
    "            z_l.append(d_stats[2])\n",
    "    print(i)\n",
    "d = {\"a\": a_l, \"b\": b_l, \"c\": c_l, \"d_stat\": dbranch_l, \"z-score\": z_l}\n",
    "df_d = pd.DataFrame(data=d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "d_stats = d_run(\"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Cynocephalus, Central Tanzania\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(-0.06475375476136486,\n",
       " 0.006005502976184223,\n",
       " -10.782403242185737,\n",
       " array([-9.55699812e-02, -1.31100120e-04, -1.17308915e-01, -9.98011877e-02,\n",
       "        -1.06624466e-01,  3.34512537e-03, -1.44540285e-01, -5.94068766e-02,\n",
       "        -7.67035446e-02, -1.12630187e-01, -1.08815703e-01, -1.29117705e-01,\n",
       "        -5.61092768e-02, -7.78683716e-02, -2.31252125e-02, -1.68808708e-02,\n",
       "        -1.06043237e-01, -1.03348065e-02, -9.12816433e-02, -6.01725900e-03,\n",
       "        -7.20214201e-02, -4.85354293e-02, -1.27845589e-01, -9.04734775e-02,\n",
       "        -1.75690147e-02, -6.92762238e-02, -2.15383849e-02, -6.20699070e-02,\n",
       "        -1.04953781e-01, -5.40716905e-02, -5.46386093e-02, -7.87172468e-02,\n",
       "        -3.42884935e-02, -3.09326065e-02, -2.15739872e-02,  1.78493232e-02,\n",
       "        -8.51412136e-02, -2.33013887e-02, -1.25808448e-01, -9.32427555e-02,\n",
       "        -1.24655501e-01, -1.56151781e-01, -2.30194356e-01, -1.54891939e-01,\n",
       "        -5.73779266e-02, -3.81763906e-02, -4.37308845e-02, -3.51838192e-02,\n",
       "        -6.02538874e-03,  6.13898147e-03, -3.75366422e-02, -1.18118528e-01,\n",
       "        -7.59592124e-02, -7.77162206e-02, -1.05679935e-01, -3.99317952e-02,\n",
       "        -7.96895752e-02, -1.22140080e-01, -3.68415400e-02, -1.53885837e-01,\n",
       "        -9.84702848e-02, -5.17583426e-02, -3.32447052e-02, -7.04540500e-02,\n",
       "        -1.33433597e-01, -2.43658921e-01, -9.84868202e-03, -4.84117552e-02,\n",
       "        -5.99622202e-03, -2.86007243e-02, -6.60008493e-02,  5.03191933e-02,\n",
       "        -9.37919172e-03, -7.69139705e-02, -5.79384233e-02, -3.84402449e-02,\n",
       "         1.60066025e-02, -7.91161522e-02, -8.60518255e-02, -1.78334988e-02]),\n",
       " array([-0.06434437, -0.06564654, -0.06402102, -0.0643182 , -0.0643285 ,\n",
       "        -0.06604077, -0.06365298, -0.06490083, -0.06466651, -0.06413579,\n",
       "        -0.06422442, -0.06376366, -0.06491845, -0.06467765, -0.0652654 ,\n",
       "        -0.06548018, -0.06432633, -0.0653538 , -0.06456012, -0.06554277,\n",
       "        -0.06468522, -0.06502657, -0.06409154, -0.06444877, -0.06546686,\n",
       "        -0.06476929, -0.06521846, -0.06484093, -0.06424856, -0.06498823,\n",
       "        -0.06495212, -0.06462692, -0.06523419, -0.06530422, -0.06538431,\n",
       "        -0.06557991, -0.06462098, -0.06521682, -0.06414288, -0.06442353,\n",
       "        -0.0640793 , -0.06384345, -0.06395775, -0.0642809 , -0.06486951,\n",
       "        -0.06502618, -0.06501995, -0.06506246, -0.06548135, -0.06569138,\n",
       "        -0.06504051, -0.06411001, -0.06463816, -0.06466679, -0.06425006,\n",
       "        -0.06510595, -0.06467032, -0.06427762, -0.06505801, -0.06402504,\n",
       "        -0.06440903, -0.06490103, -0.06518718, -0.06472976, -0.06402139,\n",
       "        -0.06216087, -0.06563477, -0.0650172 , -0.0658208 , -0.06545228,\n",
       "        -0.06479567, -0.06666148, -0.06576731, -0.06459932, -0.06493445,\n",
       "        -0.06520489, -0.06654165, -0.06462652, -0.06450126, -0.06546558]))"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "      <th>d_stat</th>\n",
       "      <th>z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.620847</td>\n",
       "      <td>-82.991862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.618753</td>\n",
       "      <td>-81.624641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.612681</td>\n",
       "      <td>-80.283408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.546236</td>\n",
       "      <td>-57.678917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.448112</td>\n",
       "      <td>-43.260072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.453201</td>\n",
       "      <td>-42.587958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.448179</td>\n",
       "      <td>-41.693126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.384539</td>\n",
       "      <td>-35.901674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.386458</td>\n",
       "      <td>-35.898824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.361958</td>\n",
       "      <td>-34.610271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.338966</td>\n",
       "      <td>-27.465743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.224969</td>\n",
       "      <td>-22.829063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.206732</td>\n",
       "      <td>-22.779319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.259809</td>\n",
       "      <td>-22.175532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.225923</td>\n",
       "      <td>-21.772632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.193661</td>\n",
       "      <td>-21.380770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.187440</td>\n",
       "      <td>-21.102967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.256815</td>\n",
       "      <td>-21.031469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.200250</td>\n",
       "      <td>-19.949935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.164108</td>\n",
       "      <td>-19.902089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.167569</td>\n",
       "      <td>-17.784128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.174307</td>\n",
       "      <td>-17.502583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.127911</td>\n",
       "      <td>-16.650792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.116889</td>\n",
       "      <td>-15.888644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.162436</td>\n",
       "      <td>-15.838126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.158227</td>\n",
       "      <td>-15.799299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.174067</td>\n",
       "      <td>-15.167331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.129338</td>\n",
       "      <td>-14.748709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.167742</td>\n",
       "      <td>-14.344351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.143480</td>\n",
       "      <td>-14.237878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>-0.111889</td>\n",
       "      <td>-14.096923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.088786</td>\n",
       "      <td>-13.587294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.059648</td>\n",
       "      <td>-13.405909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.158170</td>\n",
       "      <td>-13.206107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>-0.101757</td>\n",
       "      <td>-12.741264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.100744</td>\n",
       "      <td>-12.653245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>-0.111106</td>\n",
       "      <td>-10.662891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.059073</td>\n",
       "      <td>-10.209474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.057411</td>\n",
       "      <td>-9.249944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.066383</td>\n",
       "      <td>-8.981167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.107605</td>\n",
       "      <td>-8.714004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.087584</td>\n",
       "      <td>-6.358207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.037007</td>\n",
       "      <td>-5.079023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.066625</td>\n",
       "      <td>-5.050874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.028413</td>\n",
       "      <td>-3.964565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.020938</td>\n",
       "      <td>-3.946634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.024646</td>\n",
       "      <td>-3.099166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>-0.015598</td>\n",
       "      <td>-2.667886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.036917</td>\n",
       "      <td>-2.649355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>-0.013635</td>\n",
       "      <td>-1.988826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.011556</td>\n",
       "      <td>-1.636643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.015923</td>\n",
       "      <td>-1.210255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>-0.006956</td>\n",
       "      <td>-1.055249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.004773</td>\n",
       "      <td>-0.830046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.006034</td>\n",
       "      <td>-0.746894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>-0.003686</td>\n",
       "      <td>-0.456513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.003325</td>\n",
       "      <td>-0.345598</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  a                               b  \\\n",
       "3    Cynocephalus, Central Tanzania                Anubis, Tanzania   \n",
       "5                   Ursinus, Zambia                Anubis, Tanzania   \n",
       "6                    Kindae, Zambia                Anubis, Tanzania   \n",
       "4    Cynocephalus, Western Tanzania                Anubis, Tanzania   \n",
       "10   Cynocephalus, Central Tanzania             Hamadryas, Ethiopia   \n",
       "12                  Ursinus, Zambia             Hamadryas, Ethiopia   \n",
       "13                   Kindae, Zambia             Hamadryas, Ethiopia   \n",
       "20                   Kindae, Zambia             Hamadryas, Ethiopia   \n",
       "19                  Ursinus, Zambia             Hamadryas, Ethiopia   \n",
       "17   Cynocephalus, Central Tanzania             Hamadryas, Ethiopia   \n",
       "11   Cynocephalus, Western Tanzania             Hamadryas, Ethiopia   \n",
       "142             Hamadryas, Ethiopia                  Kindae, Zambia   \n",
       "140                Anubis, Ethiopia                  Kindae, Zambia   \n",
       "2               Hamadryas, Ethiopia                Anubis, Tanzania   \n",
       "93              Hamadryas, Ethiopia                 Ursinus, Zambia   \n",
       "141                Anubis, Tanzania                  Kindae, Zambia   \n",
       "65              Hamadryas, Ethiopia  Cynocephalus, Western Tanzania   \n",
       "18   Cynocephalus, Western Tanzania             Hamadryas, Ethiopia   \n",
       "91                 Anubis, Ethiopia                 Ursinus, Zambia   \n",
       "63                 Anubis, Ethiopia  Cynocephalus, Western Tanzania   \n",
       "92                 Anubis, Tanzania                 Ursinus, Zambia   \n",
       "55                   Kindae, Zambia  Cynocephalus, Western Tanzania   \n",
       "64                 Anubis, Tanzania  Cynocephalus, Western Tanzania   \n",
       "97                   Kindae, Zambia                 Ursinus, Zambia   \n",
       "48                   Kindae, Zambia  Cynocephalus, Western Tanzania   \n",
       "45   Cynocephalus, Central Tanzania  Cynocephalus, Western Tanzania   \n",
       "54                  Ursinus, Zambia  Cynocephalus, Western Tanzania   \n",
       "128             Hamadryas, Ethiopia                  Kindae, Zambia   \n",
       "47                  Ursinus, Zambia  Cynocephalus, Western Tanzania   \n",
       "52   Cynocephalus, Central Tanzania  Cynocephalus, Western Tanzania   \n",
       "59   Cynocephalus, Central Tanzania  Cynocephalus, Western Tanzania   \n",
       "136  Cynocephalus, Central Tanzania                  Kindae, Zambia   \n",
       "69                   Kindae, Zambia  Cynocephalus, Western Tanzania   \n",
       "135             Hamadryas, Ethiopia                  Kindae, Zambia   \n",
       "62                   Kindae, Zambia  Cynocephalus, Western Tanzania   \n",
       "126                Anubis, Ethiopia                  Kindae, Zambia   \n",
       "61                  Ursinus, Zambia  Cynocephalus, Western Tanzania   \n",
       "138                 Ursinus, Zambia                  Kindae, Zambia   \n",
       "95   Cynocephalus, Western Tanzania                 Ursinus, Zambia   \n",
       "127                Anubis, Tanzania                  Kindae, Zambia   \n",
       "100             Hamadryas, Ethiopia                 Ursinus, Zambia   \n",
       "133                Anubis, Ethiopia                  Kindae, Zambia   \n",
       "101  Cynocephalus, Central Tanzania                 Ursinus, Zambia   \n",
       "134                Anubis, Tanzania                  Kindae, Zambia   \n",
       "33                  Ursinus, Zambia  Cynocephalus, Central Tanzania   \n",
       "34                   Kindae, Zambia  Cynocephalus, Central Tanzania   \n",
       "144  Cynocephalus, Western Tanzania                  Kindae, Zambia   \n",
       "122  Cynocephalus, Central Tanzania                  Kindae, Zambia   \n",
       "98                 Anubis, Ethiopia                 Ursinus, Zambia   \n",
       "124                 Ursinus, Zambia                  Kindae, Zambia   \n",
       "110                 Ursinus, Zambia                  Kindae, Zambia   \n",
       "99                 Anubis, Tanzania                 Ursinus, Zambia   \n",
       "117                 Ursinus, Zambia                  Kindae, Zambia   \n",
       "108  Cynocephalus, Central Tanzania                  Kindae, Zambia   \n",
       "26                  Ursinus, Zambia  Cynocephalus, Central Tanzania   \n",
       "87   Cynocephalus, Central Tanzania                 Ursinus, Zambia   \n",
       "143  Cynocephalus, Central Tanzania                  Kindae, Zambia   \n",
       "\n",
       "                                  c    d_stat    z-score  \n",
       "3                  Anubis, Ethiopia -0.620847 -82.991862  \n",
       "5                  Anubis, Ethiopia -0.618753 -81.624641  \n",
       "6                  Anubis, Ethiopia -0.612681 -80.283408  \n",
       "4                  Anubis, Ethiopia -0.546236 -57.678917  \n",
       "10                 Anubis, Ethiopia -0.448112 -43.260072  \n",
       "12                 Anubis, Ethiopia -0.453201 -42.587958  \n",
       "13                 Anubis, Ethiopia -0.448179 -41.693126  \n",
       "20                 Anubis, Tanzania -0.384539 -35.901674  \n",
       "19                 Anubis, Tanzania -0.386458 -35.898824  \n",
       "17                 Anubis, Tanzania -0.361958 -34.610271  \n",
       "11                 Anubis, Ethiopia -0.338966 -27.465743  \n",
       "142                 Ursinus, Zambia -0.224969 -22.829063  \n",
       "140                 Ursinus, Zambia -0.206732 -22.779319  \n",
       "2                  Anubis, Ethiopia -0.259809 -22.175532  \n",
       "93   Cynocephalus, Central Tanzania -0.225923 -21.772632  \n",
       "141                 Ursinus, Zambia -0.193661 -21.380770  \n",
       "65   Cynocephalus, Central Tanzania -0.187440 -21.102967  \n",
       "18                 Anubis, Tanzania -0.256815 -21.031469  \n",
       "91   Cynocephalus, Central Tanzania -0.200250 -19.949935  \n",
       "63   Cynocephalus, Central Tanzania -0.164108 -19.902089  \n",
       "92   Cynocephalus, Central Tanzania -0.167569 -17.784128  \n",
       "55                 Anubis, Tanzania -0.174307 -17.502583  \n",
       "64   Cynocephalus, Central Tanzania -0.127911 -16.650792  \n",
       "97   Cynocephalus, Central Tanzania -0.116889 -15.888644  \n",
       "48                 Anubis, Ethiopia -0.162436 -15.838126  \n",
       "45                 Anubis, Ethiopia -0.158227 -15.799299  \n",
       "54                 Anubis, Tanzania -0.174067 -15.167331  \n",
       "128  Cynocephalus, Central Tanzania -0.129338 -14.748709  \n",
       "47                 Anubis, Ethiopia -0.167742 -14.344351  \n",
       "52                 Anubis, Tanzania -0.143480 -14.237878  \n",
       "59              Hamadryas, Ethiopia -0.111889 -14.096923  \n",
       "136  Cynocephalus, Western Tanzania -0.088786 -13.587294  \n",
       "69   Cynocephalus, Central Tanzania -0.059648 -13.405909  \n",
       "135  Cynocephalus, Western Tanzania -0.158170 -13.206107  \n",
       "62              Hamadryas, Ethiopia -0.101757 -12.741264  \n",
       "126  Cynocephalus, Central Tanzania -0.100744 -12.653245  \n",
       "61              Hamadryas, Ethiopia -0.111106 -10.662891  \n",
       "138  Cynocephalus, Western Tanzania -0.059073 -10.209474  \n",
       "95   Cynocephalus, Central Tanzania -0.057411  -9.249944  \n",
       "127  Cynocephalus, Central Tanzania -0.066383  -8.981167  \n",
       "100  Cynocephalus, Western Tanzania -0.107605  -8.714004  \n",
       "133  Cynocephalus, Western Tanzania -0.087584  -6.358207  \n",
       "101  Cynocephalus, Western Tanzania -0.037007  -5.079023  \n",
       "134  Cynocephalus, Western Tanzania -0.066625  -5.050874  \n",
       "33                 Anubis, Tanzania -0.028413  -3.964565  \n",
       "34                 Anubis, Tanzania -0.020938  -3.946634  \n",
       "144                 Ursinus, Zambia -0.024646  -3.099166  \n",
       "122             Hamadryas, Ethiopia -0.015598  -2.667886  \n",
       "98   Cynocephalus, Western Tanzania -0.036917  -2.649355  \n",
       "124             Hamadryas, Ethiopia -0.013635  -1.988826  \n",
       "110                Anubis, Ethiopia -0.011556  -1.636643  \n",
       "99   Cynocephalus, Western Tanzania -0.015923  -1.210255  \n",
       "117                Anubis, Tanzania -0.006956  -1.055249  \n",
       "108                Anubis, Ethiopia -0.004773  -0.830046  \n",
       "26                 Anubis, Ethiopia -0.006034  -0.746894  \n",
       "87              Hamadryas, Ethiopia -0.003686  -0.456513  \n",
       "143                 Ursinus, Zambia -0.003325  -0.345598  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_d.loc[df_d[\"z-score\"] < 0].sort_values([\"z-score\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "      <th>d_stat</th>\n",
       "      <th>z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.003686</td>\n",
       "      <td>0.456513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.006034</td>\n",
       "      <td>0.746894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.004773</td>\n",
       "      <td>0.830046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.006956</td>\n",
       "      <td>1.055249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.011556</td>\n",
       "      <td>1.636643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.013635</td>\n",
       "      <td>1.988826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.015598</td>\n",
       "      <td>2.667886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.020938</td>\n",
       "      <td>3.946634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.028413</td>\n",
       "      <td>3.964565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>0.057411</td>\n",
       "      <td>9.249944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>0.059073</td>\n",
       "      <td>10.209474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.111106</td>\n",
       "      <td>10.662891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.101757</td>\n",
       "      <td>12.741264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>0.059648</td>\n",
       "      <td>13.405909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.111889</td>\n",
       "      <td>14.096923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.143480</td>\n",
       "      <td>14.237878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.167742</td>\n",
       "      <td>14.344351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.174067</td>\n",
       "      <td>15.167331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.158227</td>\n",
       "      <td>15.799299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.162436</td>\n",
       "      <td>15.838126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>0.116889</td>\n",
       "      <td>15.888644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.174307</td>\n",
       "      <td>17.502583</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  a                               b  \\\n",
       "40                  Ursinus, Zambia  Cynocephalus, Central Tanzania   \n",
       "73   Cynocephalus, Central Tanzania                 Ursinus, Zambia   \n",
       "27                   Kindae, Zambia  Cynocephalus, Central Tanzania   \n",
       "83                   Kindae, Zambia                 Ursinus, Zambia   \n",
       "76                   Kindae, Zambia                 Ursinus, Zambia   \n",
       "90                   Kindae, Zambia                 Ursinus, Zambia   \n",
       "41                   Kindae, Zambia  Cynocephalus, Central Tanzania   \n",
       "115  Cynocephalus, Central Tanzania                  Kindae, Zambia   \n",
       "80   Cynocephalus, Central Tanzania                 Ursinus, Zambia   \n",
       "68                  Ursinus, Zambia  Cynocephalus, Western Tanzania   \n",
       "104                  Kindae, Zambia                 Ursinus, Zambia   \n",
       "88   Cynocephalus, Western Tanzania                 Ursinus, Zambia   \n",
       "123  Cynocephalus, Western Tanzania                  Kindae, Zambia   \n",
       "130  Cynocephalus, Western Tanzania                  Kindae, Zambia   \n",
       "39   Cynocephalus, Western Tanzania  Cynocephalus, Central Tanzania   \n",
       "32   Cynocephalus, Western Tanzania  Cynocephalus, Central Tanzania   \n",
       "74   Cynocephalus, Western Tanzania                 Ursinus, Zambia   \n",
       "81   Cynocephalus, Western Tanzania                 Ursinus, Zambia   \n",
       "25   Cynocephalus, Western Tanzania  Cynocephalus, Central Tanzania   \n",
       "109  Cynocephalus, Western Tanzania                  Kindae, Zambia   \n",
       "131                 Ursinus, Zambia                  Kindae, Zambia   \n",
       "116  Cynocephalus, Western Tanzania                  Kindae, Zambia   \n",
       "\n",
       "                                  c    d_stat    z-score  \n",
       "40              Hamadryas, Ethiopia  0.003686   0.456513  \n",
       "73                 Anubis, Ethiopia  0.006034   0.746894  \n",
       "27                 Anubis, Ethiopia  0.004773   0.830046  \n",
       "83                 Anubis, Tanzania  0.006956   1.055249  \n",
       "76                 Anubis, Ethiopia  0.011556   1.636643  \n",
       "90              Hamadryas, Ethiopia  0.013635   1.988826  \n",
       "41              Hamadryas, Ethiopia  0.015598   2.667886  \n",
       "115                Anubis, Tanzania  0.020938   3.946634  \n",
       "80                 Anubis, Tanzania  0.028413   3.964565  \n",
       "68   Cynocephalus, Central Tanzania  0.057411   9.249944  \n",
       "104  Cynocephalus, Western Tanzania  0.059073  10.209474  \n",
       "88              Hamadryas, Ethiopia  0.111106  10.662891  \n",
       "123             Hamadryas, Ethiopia  0.101757  12.741264  \n",
       "130  Cynocephalus, Central Tanzania  0.059648  13.405909  \n",
       "39              Hamadryas, Ethiopia  0.111889  14.096923  \n",
       "32                 Anubis, Tanzania  0.143480  14.237878  \n",
       "74                 Anubis, Ethiopia  0.167742  14.344351  \n",
       "81                 Anubis, Tanzania  0.174067  15.167331  \n",
       "25                 Anubis, Ethiopia  0.158227  15.799299  \n",
       "109                Anubis, Ethiopia  0.162436  15.838126  \n",
       "131  Cynocephalus, Central Tanzania  0.116889  15.888644  \n",
       "116                Anubis, Tanzania  0.174307  17.502583  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_d.loc[(df_d[\"z-score\"] < 20) & (df_d[\"z-score\"] > 0)].sort_values([\"z-score\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "      <th>d_stat</th>\n",
       "      <th>z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.259809</td>\n",
       "      <td>22.175532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.414274</td>\n",
       "      <td>48.325705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.620847</td>\n",
       "      <td>82.991862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.630889</td>\n",
       "      <td>98.215979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.438138</td>\n",
       "      <td>54.890754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.546236</td>\n",
       "      <td>57.678917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.571945</td>\n",
       "      <td>66.619393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.378130</td>\n",
       "      <td>42.308261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.127911</td>\n",
       "      <td>-16.650792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.618753</td>\n",
       "      <td>81.624641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.644926</td>\n",
       "      <td>101.250345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.430073</td>\n",
       "      <td>46.064819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.167569</td>\n",
       "      <td>-17.784128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.015923</td>\n",
       "      <td>-1.210255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.612681</td>\n",
       "      <td>80.283408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>0.643871</td>\n",
       "      <td>99.050193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>0.421365</td>\n",
       "      <td>50.460019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>-0.066383</td>\n",
       "      <td>-8.981167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>-0.066625</td>\n",
       "      <td>-5.050874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.193661</td>\n",
       "      <td>-21.380770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    a                               b  \\\n",
       "1    Anubis, Tanzania                Anubis, Tanzania   \n",
       "8    Anubis, Tanzania             Hamadryas, Ethiopia   \n",
       "15   Anubis, Tanzania             Hamadryas, Ethiopia   \n",
       "22   Anubis, Tanzania  Cynocephalus, Central Tanzania   \n",
       "29   Anubis, Tanzania  Cynocephalus, Central Tanzania   \n",
       "36   Anubis, Tanzania  Cynocephalus, Central Tanzania   \n",
       "43   Anubis, Tanzania  Cynocephalus, Western Tanzania   \n",
       "50   Anubis, Tanzania  Cynocephalus, Western Tanzania   \n",
       "57   Anubis, Tanzania  Cynocephalus, Western Tanzania   \n",
       "64   Anubis, Tanzania  Cynocephalus, Western Tanzania   \n",
       "71   Anubis, Tanzania                 Ursinus, Zambia   \n",
       "78   Anubis, Tanzania                 Ursinus, Zambia   \n",
       "85   Anubis, Tanzania                 Ursinus, Zambia   \n",
       "92   Anubis, Tanzania                 Ursinus, Zambia   \n",
       "99   Anubis, Tanzania                 Ursinus, Zambia   \n",
       "106  Anubis, Tanzania                  Kindae, Zambia   \n",
       "113  Anubis, Tanzania                  Kindae, Zambia   \n",
       "120  Anubis, Tanzania                  Kindae, Zambia   \n",
       "127  Anubis, Tanzania                  Kindae, Zambia   \n",
       "134  Anubis, Tanzania                  Kindae, Zambia   \n",
       "141  Anubis, Tanzania                  Kindae, Zambia   \n",
       "\n",
       "                                  c    d_stat     z-score  \n",
       "1                  Anubis, Ethiopia  0.000000         NaN  \n",
       "8                  Anubis, Ethiopia  0.259809   22.175532  \n",
       "15                 Anubis, Tanzania  0.414274   48.325705  \n",
       "22                 Anubis, Ethiopia  0.620847   82.991862  \n",
       "29                 Anubis, Tanzania  0.630889   98.215979  \n",
       "36              Hamadryas, Ethiopia  0.438138   54.890754  \n",
       "43                 Anubis, Ethiopia  0.546236   57.678917  \n",
       "50                 Anubis, Tanzania  0.571945   66.619393  \n",
       "57              Hamadryas, Ethiopia  0.378130   42.308261  \n",
       "64   Cynocephalus, Central Tanzania -0.127911  -16.650792  \n",
       "71                 Anubis, Ethiopia  0.618753   81.624641  \n",
       "78                 Anubis, Tanzania  0.644926  101.250345  \n",
       "85              Hamadryas, Ethiopia  0.430073   46.064819  \n",
       "92   Cynocephalus, Central Tanzania -0.167569  -17.784128  \n",
       "99   Cynocephalus, Western Tanzania -0.015923   -1.210255  \n",
       "106                Anubis, Ethiopia  0.612681   80.283408  \n",
       "113                Anubis, Tanzania  0.643871   99.050193  \n",
       "120             Hamadryas, Ethiopia  0.421365   50.460019  \n",
       "127  Cynocephalus, Central Tanzania -0.066383   -8.981167  \n",
       "134  Cynocephalus, Western Tanzania -0.066625   -5.050874  \n",
       "141                 Ursinus, Zambia -0.193661  -21.380770  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_d.loc[(df_d.a == \"Anubis, Tanzania\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>a</th>\n",
       "      <th>b</th>\n",
       "      <th>c</th>\n",
       "      <th>d_stat</th>\n",
       "      <th>z-score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.308653</td>\n",
       "      <td>47.035724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.259809</td>\n",
       "      <td>-22.175532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.620847</td>\n",
       "      <td>-82.991862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Anubis, Tanzania</td>\n",
       "      <td>Anubis, Ethiopia</td>\n",
       "      <td>-0.546236</td>\n",
       "      <td>-57.678917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>Hamadryas, Ethiopia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.224969</td>\n",
       "      <td>-22.829063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>Cynocephalus, Central Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.003325</td>\n",
       "      <td>-0.345598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Cynocephalus, Western Tanzania</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>-0.024646</td>\n",
       "      <td>-3.099166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>0.616927</td>\n",
       "      <td>66.007375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Kindae, Zambia</td>\n",
       "      <td>Ursinus, Zambia</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>147 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  a                 b                 c  \\\n",
       "0                  Anubis, Ethiopia  Anubis, Tanzania  Anubis, Ethiopia   \n",
       "1                  Anubis, Tanzania  Anubis, Tanzania  Anubis, Ethiopia   \n",
       "2               Hamadryas, Ethiopia  Anubis, Tanzania  Anubis, Ethiopia   \n",
       "3    Cynocephalus, Central Tanzania  Anubis, Tanzania  Anubis, Ethiopia   \n",
       "4    Cynocephalus, Western Tanzania  Anubis, Tanzania  Anubis, Ethiopia   \n",
       "..                              ...               ...               ...   \n",
       "142             Hamadryas, Ethiopia    Kindae, Zambia   Ursinus, Zambia   \n",
       "143  Cynocephalus, Central Tanzania    Kindae, Zambia   Ursinus, Zambia   \n",
       "144  Cynocephalus, Western Tanzania    Kindae, Zambia   Ursinus, Zambia   \n",
       "145                 Ursinus, Zambia    Kindae, Zambia   Ursinus, Zambia   \n",
       "146                  Kindae, Zambia    Kindae, Zambia   Ursinus, Zambia   \n",
       "\n",
       "       d_stat    z-score  \n",
       "0    0.308653  47.035724  \n",
       "1    0.000000        NaN  \n",
       "2   -0.259809 -22.175532  \n",
       "3   -0.620847 -82.991862  \n",
       "4   -0.546236 -57.678917  \n",
       "..        ...        ...  \n",
       "142 -0.224969 -22.829063  \n",
       "143 -0.003325  -0.345598  \n",
       "144 -0.024646  -3.099166  \n",
       "145  0.616927  66.007375  \n",
       "146  0.000000        NaN  \n",
       "\n",
       "[147 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def f_hat_d(aca, acb, acc, acd):\n",
    "    a = aca.to_frequencies()[:, 1]\n",
    "    b = acb.to_frequencies()[:, 1]\n",
    "    c = acc.to_frequencies()[:, 1]\n",
    "    d = acd.to_frequencies()[:, 1]\n",
    "    num = (a - b) * (c - d)\n",
    "    max_bc = np.maximum(b, c)\n",
    "    den = (a + max_bc - (2 * a * max_bc)) * (max_bc + d - (2 * max_bc * d))\n",
    "\n",
    "    return num, den\n",
    "\n",
    "def average_f_hat_d(aca, acb, acc, acd, blen):\n",
    "    from allel.stats.window import moving_statistic\n",
    "    from allel.stats.misc import jackknife\n",
    "    num, den = f_hat_d(aca, acb, acc, acd)\n",
    "\n",
    "    # N.B., nans can occur if any of the populations have completely missing\n",
    "    # genotype calls at a variant (i.e., allele number is zero). Here we\n",
    "    # assume that is rare enough to be negligible.\n",
    "\n",
    "    # calculate overall estimate\n",
    "    d_avg = np.nansum(num) / np.nansum(den)\n",
    "\n",
    "    # compute the numerator and denominator within each block\n",
    "    num_bsum = moving_statistic(num, statistic=np.nansum, size=blen)\n",
    "    den_bsum = moving_statistic(den, statistic=np.nansum, size=blen)\n",
    "\n",
    "    # calculate the statistic values in each block\n",
    "    vb = num_bsum / den_bsum\n",
    "\n",
    "    # estimate standard error\n",
    "    _, se, vj = jackknife((num_bsum, den_bsum),\n",
    "                          statistic=lambda n, d: np.sum(n) / np.sum(d))\n",
    "\n",
    "    # compute Z score\n",
    "    z = d_avg / se\n",
    "\n",
    "    return d_avg, se, z, vb, vj\n",
    "\n",
    "def d_stats_trios(origin_l_l, chrom_list):\n",
    "    a_l = []\n",
    "    b_l = []\n",
    "    c_l = []\n",
    "    dbranch_l = []\n",
    "    z_l = []\n",
    "    chrom_l = []\n",
    "    se_l = []\n",
    "    for chrom in chrom_list:\n",
    "        print(\"Working on chr\", chrom)\n",
    "        zarr_dir = \"/faststorage/project/baboondiversity/data/PG_baboons_mmul_20_04_2021/callset.zarr/chr{}\"\n",
    "        callset = zarr.open_group(zarr_dir.format(chrom), mode=\"r\")\n",
    "        gt = allel.GenotypeArray(callset[\"calldata/GT\"])\n",
    "        bi_sites = gt.count_alleles().is_biallelic()\n",
    "        pos = callset[\"variants/POS\"][:].compress(bi_sites)\n",
    "        pos_sorted = allel.SortedIndex(pos)\n",
    "        gt = gt.compress(bi_sites)\n",
    "        d_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == \"Gelada, Captive\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "        for origin_trio in origin_l_l:\n",
    "            a, b, c = origin_trio[0], origin_trio[1], origin_trio[2]\n",
    "            a_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == a) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "            b_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == b) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "            c_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == c) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "            d_stats = average_f_hat_d(a_ac, b_ac, c_ac, d_ac, blen=100000)\n",
    "            a_l.append(a)\n",
    "            b_l.append(b)\n",
    "            c_l.append(c)\n",
    "            dbranch_l.append(-d_stats[0])\n",
    "            se_l.append(d_stats[1])\n",
    "            z_l.append(d_stats[2])\n",
    "            chrom_l.append(\"chr\"+chrom)\n",
    "    d = {\"a\": a_l, \"b\": b_l, \"c\": c_l, \"d_stat\": dbranch_l, \"z-score\": z_l, \"SE\": se_l, \"chrom\": chrom_l}\n",
    "    df_d = pd.DataFrame(data=d)\n",
    "    return df_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'd_stats_trios' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-f0e9bab47624>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m               [\"Anubis, Tanzania\", \"Anubis, Ethiopia\", \"Hamadryas, Ethiopia\"]]\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0md_stats_trios\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0md_stats_trios\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0morigin_l_l\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mchromosome_numbers\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'd_stats_trios' is not defined"
     ]
    }
   ],
   "source": [
    "chromosome_numbers = ['{}'.format(x) for x in range(1, 21)] + ['X']\n",
    "origin_l_l = [[\"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Cynocephalus, Central Tanzania\"],\n",
    "              [\"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Kindae, Zambia\"],\n",
    "              [\"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Cynocephalus, Western Tanzania\"],\n",
    "              [\"Anubis, Tanzania\", \"Anubis, Ethiopia\", \"Hamadryas, Ethiopia\"]]\n",
    "\n",
    "d_stats_trios = d_stats_trios(origin_l_l, chromosome_numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d_stats_trios.loc[d_stats_trios[\"chrom\"] == \"chrX\", \"chrtype\"] = \"chrX\"\n",
    "d_stats_trios.loc[d_stats_trios[\"chrom\"] != \"chrX\", \"chrtype\"] = \"autosomes\"\n",
    "d_stats_trios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d_stats_trios.loc[(d_stats_trios.c ==\"Hamadryas, Ethiopia\")].groupby([\"chrtype\"]).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d_stats_trios.loc[(d_stats_trios.c ==\"Cynocephalus, Central Tanzania\")].groupby([\"chrtype\"]).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.scatterplot(data=d_stats_trios.loc[(d_stats_trios.c ==\"Hamadryas, Ethiopia\")], x=\"d_stat\", y=\"SE\", hue=\"chrtype\")\n",
    "fig = g.get_figure()\n",
    "g.legend(bbox_to_anchor=(1., 0.5, 0., 0))\n",
    "g.set(title=\"Hamadryas to Ethiopian Anubis\", xlabel=\"f_hat\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(d_stats_trios.loc[(d_stats_trios.c ==\"Hamadryas, Ethiopia\") & (d_stats_trios.chrom != \"chrX\")].mean(),\n",
    "d_stats_trios.loc[(d_stats_trios.c ==\"Hamadryas, Ethiopia\") & (d_stats_trios.chrom == \"chrX\")].mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for trio in origin_l_l:\n",
    "    pure_df = d_stats_trios.loc[(d_stats_trios.b == trio[1]) & (d_stats_trios.c == trio[2])]\n",
    "    print(trio)\n",
    "    print(pure_df.loc[pure_df.chrtype == \"chrX\"].d_stat.mean(), pure_df.loc[pure_df.chrtype == \"chrX\"][\"z-score\"].mean())\n",
    "    print(pure_df.loc[pure_df.chrtype != \"chrX\"].d_stat.mean(), pure_df.loc[pure_df.chrtype != \"chrX\"][\"z-score\"].mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These results are quite close to the papio ref results. I will use the At/Ae/H and Ae/At/Cct as the two admixture cases, so I ignore the kindae and western tanzania cynocephalus.\n",
    "I will calculate all statistics using the windowed functions in scikit-allel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#Anubis, Tanzania - Anubis, Ethiopia - Hamadryas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Various functions to use in calculate the stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bisect\n",
    "def f_hat_d(aca, acb, acc, acd):\n",
    "    a = aca.to_frequencies()[:, 1]\n",
    "    b = acb.to_frequencies()[:, 1]\n",
    "    c = acc.to_frequencies()[:, 1]\n",
    "    d = acd.to_frequencies()[:, 1]\n",
    "    num = (a - b) * (c - d)\n",
    "    max_bc = np.maximum(b, c)\n",
    "    den = (a + max_bc - (2 * a * max_bc)) * (max_bc + d - (2 * max_bc * d))\n",
    "\n",
    "    return num, den\n",
    "\n",
    "def locate_range_empty_allowed(self, start=None, stop=None):\n",
    "        if start is None:\n",
    "            start_index = 0\n",
    "        else:\n",
    "            start_index = bisect.bisect_left(self, start)\n",
    "        if stop is None:\n",
    "            stop_index = len(self)\n",
    "        else:\n",
    "            stop_index = bisect.bisect_right(self, stop)\n",
    "\n",
    "        loc = slice(start_index, stop_index)\n",
    "        return loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bi_sites = gt.count_alleles().is_biallelic()\n",
    "pos = callset[\"variants/POS\"][:].compress(bi_sites)\n",
    "gt = gt.compress(bi_sites)\n",
    "g_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == \"Gelada, Captive\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "at_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == \"Anubis, Tanzania\") &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "ae_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == \"Anubis, Ethiopia\") &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "h_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == \"Hamadryas, Ethiopia\") &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Small test of f_hat and D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f_hat_est = f_hat_d(at_ac, ae_ac, h_ac, g_ac)\n",
    "d_est = allel.patterson_d(at_ac, ae_ac, h_ac, g_ac)\n",
    "print(np.nansum(f_hat_est[0])/np.nansum(f_hat_est[1]), np.nansum(d_est[0])/np.nansum(d_est[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generating 3 estimates of admixture: Pattersons D, f_hat and shared_percentage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def windowed_stats_chrom(a, b, c, d, pos, s):\n",
    "    pos_sorted = allel.SortedIndex(pos)\n",
    "    start_l = []\n",
    "    n_s_l = []\n",
    "    d_l = []\n",
    "    f_hat_l = []\n",
    "    a_count = []\n",
    "    c_count = []\n",
    "    d_est = allel.patterson_d(a, b, c, d) #Calculating Pattersons D for all sites\n",
    "    f_hat = f_hat_d(a, b, c, d) #Calculating f_hat for all sites\n",
    "    for i in range(0, pos[-1], s):\n",
    "        pos_range = locate_range_empty_allowed(pos_sorted, i, i+s-1)\n",
    "        pos_window = pos_sorted[pos_range]\n",
    "        start_l.append(i)\n",
    "        n_s_l.append(len(pos_window))\n",
    "        if len(pos_window) > 10:\n",
    "            d_l.append(np.nansum(d_est[0][pos_range])/np.nansum(d_est[1][pos_range]))\n",
    "            f_hat_l.append(np.nansum(f_hat[0][pos_range])/np.nansum(f_hat[1][pos_range]))\n",
    "            a_sub = (a[pos_range]).to_frequencies()[:, 1]\n",
    "            b_sub = (b[pos_range]).to_frequencies()[:, 1]\n",
    "            c_sub = (c[pos_range]).to_frequencies()[:, 1]\n",
    "            case_a = (a_sub >= 0.2) & (c_sub == 0)\n",
    "            case_c = (a_sub == 0.0) & (c_sub >= 0.2)\n",
    "            a_count.append(np.nansum(b_sub.compress(case_a)))\n",
    "            c_count.append(np.nansum(b_sub.compress(case_c)))\n",
    "        else:\n",
    "            d_l.append(np.nan)\n",
    "            f_hat_l.append(np.nan)\n",
    "            a_count.append(0)\n",
    "            c_count.append(0)\n",
    "    df = pd.DataFrame(data={\"start\": start_l, \"sites\": n_s_l, \"D\": d_l, \"f_hat\": f_hat_l,\n",
    "                            \"a_shared\": a_count, \"c_shared\": c_count})\n",
    "    df[\"shared_sum_freq\"] = (df[\"c_shared\"]+df[\"a_shared\"])\n",
    "    df[\"c_percentage\"] = df[\"c_shared\"]/df[\"shared_sum_freq\"]\n",
    "    return df\n",
    "\n",
    "#chrom_df = windowed_stats_chrom(at_ac, ae_ac, h_ac, g_ac, pos, 1000000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "F_hat is more conservative than D. It always outputs an equal or less extreme statistic (this is all calculated with a sign inversion)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculating LD the recipient in the windows - higher LD in admixed windows can dispel theories about ILS or noise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ae_gt = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == \"Anubis, Ethiopia\") &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ae_n_alt = ae_gt.to_n_alt()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bool_list = []\n",
    "for i in range(ae_n_alt.shape[0]):\n",
    "    if np.all(ae_n_alt[i] == ae_n_alt[i][0]):\n",
    "        bool_list.append(False)\n",
    "    else:\n",
    "        bool_list.append(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "informative_sites = ae_n_alt.compress(bool_list, axis=0)\n",
    "informative_sites_pos = pos.compress(bool_list, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nnz, windows, counts = allel.windowed_r_squared(informative_sites_pos, informative_sites, start=0, size=100000, percentile=70)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(counts, nnz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def windowed_ld(gt, pos, size):\n",
    "    n_alt = gt.to_n_alt()\n",
    "    bool_list = []\n",
    "    for i in range(n_alt.shape[0]):\n",
    "        if np.all(n_alt[i] == n_alt[i][0]):\n",
    "            bool_list.append(False)\n",
    "        else:\n",
    "            bool_list.append(True)\n",
    "    informative_sites = n_alt.compress(bool_list, axis=0)\n",
    "    informative_sites_pos = pos.compress(bool_list, axis=0)\n",
    "    nnz, windows, counts = allel.windowed_r_squared(informative_sites_pos, informative_sites, start=0, size=size, percentile=70)\n",
    "    return pd.DataFrame(data={\"LD\": nnz, \"LD_counts\": counts})\n",
    "#windowed_ld(ae_gt, pos, 100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_input(a_name, b_name, c_name, d_name, chrom_number):\n",
    "    zarr_dir = \"/faststorage/project/baboondiversity/data/PG_baboons_mmul_20_04_2021/callset.zarr/chr{}\"\n",
    "    callset = zarr.open_group(zarr_dir.format(chrom_number), mode=\"r\")\n",
    "    gt = allel.GenotypeArray(callset[\"calldata/GT\"])\n",
    "    bi_sites = gt.count_alleles().is_biallelic()\n",
    "    pos = callset[\"variants/POS\"][:].compress(bi_sites)\n",
    "    gt = gt.compress(bi_sites)\n",
    "    a_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == a_name) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "    b_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == b_name) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "    c_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == c_name) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "    d_ac = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == d_name)].callset_index.values, axis=1).count_alleles(max_allele=1)\n",
    "    b_gt = gt.take(meta_data_samples.loc[(meta_data_samples.C_origin == b_name) &\n",
    "                                    (meta_data_samples.Sex == \"F\")].callset_index.values, axis=1)\n",
    "    return a_ac, b_ac, c_ac, d_ac, b_gt, pos"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Setting it all up to run for multiple chromosomes or multiple input combinations.\n",
    "Starting input: Names of the 4 C_origin to use, as well as chromosome number and size of windows.\n",
    "Overview of output generated by generate_input:\n",
    "Filter for biallelic sites, generate pos, allele counts for each node, gt for recipient species.\n",
    "As well, specify window size and chromosomes to investigate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Case1: Flow from hamadryas to anubis in ethiopia\n",
    "a_name, b_name, c_name, d_name = \"Anubis, Tanzania\", \"Anubis, Ethiopia\", \"Hamadryas, Ethiopia\", \"Gelada, Captive\"\n",
    "chromosome_numbers = ['{}'.format(x) for x in range(1, 21)] + ['X']\n",
    "size = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = []\n",
    "for chrom_number in chromosome_numbers:\n",
    "    a_ac, b_ac, c_ac, d_ac, b_gt, pos = generate_input(a_name, b_name, c_name, d_name, chrom_number)\n",
    "    window_df = windowed_stats_chrom(a_ac, b_ac, c_ac, d_ac, pos, size)\n",
    "    #ld_df = windowed_ld(b_gt, pos, size)\n",
    "    chrom_df = window_df #Tempoary change to not run LD analysis, as it is slow\n",
    "    #chrom_df = pd.concat([window_df, ld_df], axis=1)\n",
    "    chrom_df.insert(loc=2, column=\"chrom\", value=\"chr\"+chrom_number)\n",
    "    df_list.append(chrom_df)\n",
    "    print(chrom_number)\n",
    "aah_df = pd.concat(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Case2, gene flow from cynocephalus in central tanzania to anubis\n",
    "a_name, b_name, c_name, d_name = \"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Cynocephalus, Central Tanzania\", \"Gelada, Captive\"\n",
    "chromosome_numbers = ['{}'.format(x) for x in range(1, 21)] + ['X']\n",
    "size = 1000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_list = []\n",
    "for chrom_number in chromosome_numbers:\n",
    "    a_ac, b_ac, c_ac, d_ac, b_gt, pos = generate_input(a_name, b_name, c_name, d_name, chrom_number)\n",
    "    window_df = windowed_stats_chrom(a_ac, b_ac, c_ac, d_ac, pos, size)\n",
    "    #ld_df = windowed_ld(b_gt, pos, size)\n",
    "    chrom_df = window_df #Tempoary change to not run LD analysis, as it is slow\n",
    "    #chrom_df = pd.concat([window_df, ld_df], axis=1)\n",
    "    chrom_df.insert(loc=2, column=\"chrom\", value=\"chr\"+chrom_number)\n",
    "    df_list.append(chrom_df)\n",
    "    print(chrom_number)\n",
    "aac_df = pd.concat(df_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look at the two dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aah_df.dropna(inplace=True)\n",
    "aah_df.loc[aah_df[\"chrom\"] == \"chrX\", \"chrtype\"] = \"chrX\"\n",
    "aah_df.loc[aah_df[\"chrom\"] != \"chrX\", \"chrtype\"] = \"autosome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aah_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(data=aah_df, x=\"f_hat\", hue=\"chrtype\", common_norm=False, stat=\"probability\", multiple=\"dodge\", bins=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(data=aah_df, x=\"sites\", y=\"D\", hue=\"chrtype\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(data=aah_df, x=\"f_hat\", y=\"D\", hue=\"chrtype\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aah_df.loc[aah_df.f_hat < -0.25]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aah_df.loc[aah_df.chrom == \"chrX\"].f_hat.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aah_df.loc[aah_df.chrom != \"chrX\"].f_hat.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_df.dropna(inplace=True)\n",
    "aac_df.loc[aac_df[\"chrom\"] == \"chrX\", \"chrtype\"] = \"chrX\"\n",
    "aac_df.loc[aac_df[\"chrom\"] != \"chrX\", \"chrtype\"] = \"autosome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(data=aac_df.loc[(aac_df.f_hat > -0.3) & ], x=\"f_hat\", hue=\"chrtype\", common_norm=False, stat=\"probability\", multiple=\"dodge\", bins=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.scatterplot(data=aac_df, x=\"f_hat\", y=\"D\", hue=\"chrtype\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_df.loc[aac_df.f_hat < -0.13]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_df.loc[aac_df.chrom == \"chrX\"].f_hat.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_df.loc[aac_df.chrom != \"chrX\"].f_hat.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_df.loc[(aac_df.chrom == \"chrX\") & (aac_df.LD > 0.5) & (aac_df.f_hat < -0.05)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.corrcoef(aac_df.loc[aac_df.chrom != \"chrX\"].f_hat, aac_df.loc[aac_df.chrom != \"chrX\"].LD)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generating D with jackknife for all chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_name, b_name, c_name, d_name = \"Anubis, Tanzania\", \"Anubis, Ethiopia\", \"Hamadryas, Ethiopia\", \"Gelada, Captive\"\n",
    "chromosome_numbers = ['{}'.format(x) for x in range(1, 21)] + ['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_list = []\n",
    "d_list = []\n",
    "z_list = []\n",
    "for chrom_number in chromosome_numbers:\n",
    "    a_ac, b_ac, c_ac, d_ac, b_gt, pos = generate_input(a_name, b_name, c_name, d_name, chrom_number)\n",
    "    d, se, z, vb, vj = allel.average_patterson_d(a_ac, b_ac, c_ac, d_ac, 10000)\n",
    "    c_list.append(\"chr\"+chrom_number), d_list.append(d), z_list.append(z)\n",
    "    print(chrom_number)\n",
    "aah_jack_df = pd.DataFrame(data={\"chrom\": c_list, \"D\": d_list, \"Z\": z_list})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aah_jack_df.loc[aah_jack_df[\"chrom\"] == \"chrX\", \"chrtype\"] = \"chrX\"\n",
    "aah_jack_df.loc[aah_jack_df[\"chrom\"] != \"chrX\", \"chrtype\"] = \"autosome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.scatterplot(data=aah_jack_df, x=\"D\", y=\"Z\", hue=\"chrtype\")\n",
    "g.set_title(\"Anubis-Hamadryas case\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a_name, b_name, c_name, d_name = \"Anubis, Ethiopia\", \"Anubis, Tanzania\", \"Cynocephalus, Central Tanzania\", \"Gelada, Captive\"\n",
    "chromosome_numbers = ['{}'.format(x) for x in range(1, 21)] + ['X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_list = []\n",
    "d_list = []\n",
    "z_list = []\n",
    "for chrom_number in chromosome_numbers:\n",
    "    a_ac, b_ac, c_ac, d_ac, b_gt, pos = generate_input(a_name, b_name, c_name, d_name, chrom_number)\n",
    "    d, se, z, vb, vj = allel.average_patterson_d(a_ac, b_ac, c_ac, d_ac, 10000)\n",
    "    c_list.append(\"chr\"+chrom_number), d_list.append(d), z_list.append(z)\n",
    "    print(chrom_number)\n",
    "aac_jack_df = pd.DataFrame(data={\"chrom\": c_list, \"D\": d_list, \"Z\": z_list})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_jack_df.loc[aac_jack_df[\"chrom\"] == \"chrX\", \"chrtype\"] = \"chrX\"\n",
    "aac_jack_df.loc[aac_jack_df[\"chrom\"] != \"chrX\", \"chrtype\"] = \"autosome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.scatterplot(data=aac_jack_df, x=\"D\", y=\"Z\", hue=\"chrtype\")\n",
    "g.set_title(\"Anubis-Cynocephalus case\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aac_jack_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
