{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, I will look at the various files generated by the fineSTRUCTURE run on chr8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run ../scripts/notebook_settings.py\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['savefig.dpi'] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkcounts_path = \"/home/eriks/baboondiversity/people/eriks/baboon_first_analysis/steps/fs/chr8_run/chr8_run_linked.chunkcounts.out\"\n",
    "mcmcmatrix_path = \"/home/eriks/baboondiversity/people/eriks/baboon_first_analysis/steps/fs/chr8_run/mcmcmatrix_chr8.txt\"\n",
    "metadata_path = \"../data/Papio_metadata_with_clustering.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkcounts = pd.read_csv(chunkcounts_path, sep=\" \", header = 1, index_col=0)\n",
    "mcmcmatrix = pd.read_csv(mcmcmatrix_path, sep=\",\", index_col=0)\n",
    "meta_data_samples = pd.read_csv(metadata_path, sep =\" \")\n",
    "meta_data_samples_Sci = meta_data_samples.copy()\n",
    "for i, row in meta_data_samples_Sci.iterrows():\n",
    "    if row.PGDP_ID[0] != \"P\":\n",
    "        meta_data_samples_Sci.at[i, \"PGDP_ID\"] = \"Sci_\"+str(row.PGDP_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determining clusters by considering dendrogram, PCA, heatmap and sampling location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data_samples_Sci.index = meta_data_samples_Sci.PGDP_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "kindae = meta_data_samples_Sci.reindex(mcmcmatrix.index)[:27].C_origin\n",
    "cyno_western = meta_data_samples_Sci.reindex(mcmcmatrix.index)[27:37].C_origin\n",
    "ursinus = meta_data_samples_Sci.reindex(mcmcmatrix.index)[37:41].C_origin\n",
    "cyno_mikumi = meta_data_samples_Sci.reindex(mcmcmatrix.index)[41:78].Origin\n",
    "cyno_central = meta_data_samples_Sci.reindex(mcmcmatrix.index)[78:93].Origin #1 Mikumi is also in here\n",
    "anubis_gombe = meta_data_samples_Sci.reindex(mcmcmatrix.index)[93:110].Origin\n",
    "anubis_kenya = meta_data_samples_Sci.reindex(mcmcmatrix.index)[110:112].Origin\n",
    "anubis_serengeti = meta_data_samples_Sci.reindex(mcmcmatrix.index)[112:126].Origin\n",
    "anubis_manyara_ngorongoro = meta_data_samples_Sci.reindex(mcmcmatrix.index)[126:151].Origin\n",
    "anubis_arusha_tarangire = meta_data_samples_Sci.reindex(mcmcmatrix.index)[151:162].Origin\n",
    "anubis_ethiopia = meta_data_samples_Sci.reindex(mcmcmatrix.index)[162:187].Origin\n",
    "gelada_captive = meta_data_samples_Sci.reindex(mcmcmatrix.index)[187:189].Origin\n",
    "papio_senegal = meta_data_samples_Sci.reindex(mcmcmatrix.index)[189:201].Origin\n",
    "hamadryas_ethiopia = meta_data_samples_Sci.reindex(mcmcmatrix.index)[201:].Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df = pd.DataFrame()\n",
    "pop_df[\"ID\"] = meta_data_samples_Sci.reindex(mcmcmatrix.index).PGDP_ID\n",
    "pop_df[\"population\"] = \"unknown\"\n",
    "pop_df[\"inclusion\"] = \"unknown\"\n",
    "pop_df.iloc[0:27,1], pop_df.iloc[0:27,2] = \"kindae\", 1\n",
    "pop_df.iloc[27:37,1], pop_df.iloc[27:37,2] = \"cyno_western\", 1\n",
    "pop_df.iloc[37:41,1], pop_df.iloc[37:41,2] = \"ursinus\", 1\n",
    "pop_df.iloc[41:78,1], pop_df.iloc[41:78,2] = \"cyno_mikumi\", 1\n",
    "pop_df.iloc[78:93,1], pop_df.iloc[78:93,2] = \"cyno_RS\", 1\n",
    "pop_df.iloc[93:110,1], pop_df.iloc[93:110,2] = \"anubis_gombe\", 1\n",
    "pop_df.iloc[110:112,1], pop_df.iloc[110:112,2] = \"anubis_kenya\", 0\n",
    "pop_df.iloc[112:126,1], pop_df.iloc[112:126,2] = \"anubis_serengeti\", 1\n",
    "pop_df.iloc[126:162,1], pop_df.iloc[126:162,2] = \"anubis_MGAT\", 1\n",
    "pop_df.iloc[162:187,1], pop_df.iloc[162:187,2] = \"anubis_ethiopia\", 1\n",
    "pop_df.iloc[187:189,1], pop_df.iloc[187:189,2] = \"gelada\", 0\n",
    "pop_df.iloc[189:201,1], pop_df.iloc[189:201,2] = \"papio\", 1\n",
    "pop_df.iloc[201:,1], pop_df.iloc[201:,2] = \"hamadryas\", 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df = pop_df.reindex(meta_data_samples_Sci.index)\n",
    "pop_df.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/idfile_fs_cluster.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subset for only the females in the chrX vcf. Alan has followed my gendering, so that is nice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_x = \"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/chrX/chrX.phased.rehead.vcf.gz\"\n",
    "callset = allel.read_vcf(vcf_x)\n",
    "pop_df_females = pop_df.reindex(callset[\"samples\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_females.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/chrX_females.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>population</th>\n",
       "      <th>inclusion</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGDP_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sci_16066</th>\n",
       "      <td>Sci_16066</td>\n",
       "      <td>cyno_mikumi</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sci_16098</th>\n",
       "      <td>Sci_16098</td>\n",
       "      <td>cyno_mikumi</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sci_30977</th>\n",
       "      <td>Sci_30977</td>\n",
       "      <td>anubis_kenya</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sci_34449</th>\n",
       "      <td>Sci_34449</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sci_34474</th>\n",
       "      <td>Sci_34474</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD_0788</th>\n",
       "      <td>PD_0788</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD_0789</th>\n",
       "      <td>PD_0789</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD_0790</th>\n",
       "      <td>PD_0790</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD_0791</th>\n",
       "      <td>PD_0791</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD_0792</th>\n",
       "      <td>PD_0792</td>\n",
       "      <td>kindae</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ID    population inclusion\n",
       "PGDP_ID                                     \n",
       "Sci_16066  Sci_16066   cyno_mikumi         1\n",
       "Sci_16098  Sci_16098   cyno_mikumi         1\n",
       "Sci_30977  Sci_30977  anubis_kenya         0\n",
       "Sci_34449  Sci_34449        kindae         1\n",
       "Sci_34474  Sci_34474        kindae         1\n",
       "...              ...           ...       ...\n",
       "PD_0788      PD_0788        kindae         1\n",
       "PD_0789      PD_0789        kindae         1\n",
       "PD_0790      PD_0790        kindae         1\n",
       "PD_0791      PD_0791        kindae         1\n",
       "PD_0792      PD_0792        kindae         1\n",
       "\n",
       "[99 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pop_df_females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sci_16066</th>\n",
       "      <th>Sci_16098</th>\n",
       "      <th>Sci_30877</th>\n",
       "      <th>Sci_30977</th>\n",
       "      <th>Sci_34449</th>\n",
       "      <th>Sci_34474</th>\n",
       "      <th>Sci_38168</th>\n",
       "      <th>Sci_97074</th>\n",
       "      <th>Sci_97124</th>\n",
       "      <th>PD_0067</th>\n",
       "      <th>...</th>\n",
       "      <th>PD_0785</th>\n",
       "      <th>PD_0786</th>\n",
       "      <th>PD_0787</th>\n",
       "      <th>PD_0788</th>\n",
       "      <th>PD_0789</th>\n",
       "      <th>PD_0790</th>\n",
       "      <th>PD_0791</th>\n",
       "      <th>PD_0792</th>\n",
       "      <th>PD_0793</th>\n",
       "      <th>PD_0794_BAB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Recipient</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Sci_38168</th>\n",
       "      <td>13.168770</td>\n",
       "      <td>15.843611</td>\n",
       "      <td>9.438110</td>\n",
       "      <td>10.618792</td>\n",
       "      <td>21.212816</td>\n",
       "      <td>21.165529</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>10.915043</td>\n",
       "      <td>12.595547</td>\n",
       "      <td>4456.85035</td>\n",
       "      <td>...</td>\n",
       "      <td>14.772215</td>\n",
       "      <td>21.242087</td>\n",
       "      <td>19.059621</td>\n",
       "      <td>16.579871</td>\n",
       "      <td>17.214142</td>\n",
       "      <td>21.097534</td>\n",
       "      <td>15.314339</td>\n",
       "      <td>19.649880</td>\n",
       "      <td>16.468717</td>\n",
       "      <td>18.703289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PD_0067</th>\n",
       "      <td>13.999878</td>\n",
       "      <td>13.318741</td>\n",
       "      <td>9.699941</td>\n",
       "      <td>10.741143</td>\n",
       "      <td>13.442803</td>\n",
       "      <td>15.844821</td>\n",
       "      <td>5194.56355</td>\n",
       "      <td>9.195348</td>\n",
       "      <td>9.391233</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>15.253415</td>\n",
       "      <td>18.517194</td>\n",
       "      <td>17.025187</td>\n",
       "      <td>12.827421</td>\n",
       "      <td>16.295477</td>\n",
       "      <td>16.044948</td>\n",
       "      <td>21.018025</td>\n",
       "      <td>17.076559</td>\n",
       "      <td>16.641086</td>\n",
       "      <td>14.036267</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 227 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sci_16066  Sci_16098  Sci_30877  Sci_30977  Sci_34449  Sci_34474  \\\n",
       "Recipient                                                                     \n",
       "Sci_38168  13.168770  15.843611   9.438110  10.618792  21.212816  21.165529   \n",
       "PD_0067    13.999878  13.318741   9.699941  10.741143  13.442803  15.844821   \n",
       "\n",
       "            Sci_38168  Sci_97074  Sci_97124     PD_0067  ...    PD_0785  \\\n",
       "Recipient                                                ...              \n",
       "Sci_38168     0.00000  10.915043  12.595547  4456.85035  ...  14.772215   \n",
       "PD_0067    5194.56355   9.195348   9.391233     0.00000  ...  15.253415   \n",
       "\n",
       "             PD_0786    PD_0787    PD_0788    PD_0789    PD_0790    PD_0791  \\\n",
       "Recipient                                                                     \n",
       "Sci_38168  21.242087  19.059621  16.579871  17.214142  21.097534  15.314339   \n",
       "PD_0067    18.517194  17.025187  12.827421  16.295477  16.044948  21.018025   \n",
       "\n",
       "             PD_0792    PD_0793  PD_0794_BAB  \n",
       "Recipient                                     \n",
       "Sci_38168  19.649880  16.468717    18.703289  \n",
       "PD_0067    17.076559  16.641086    14.036267  \n",
       "\n",
       "[2 rows x 227 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chunkcounts.loc[chunkcounts.max() > 4000]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'PGDP_ID' is both an index level and a column label, which is ambiguous.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-69dceb98e026>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0msorted_chunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mchunkcounts\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted_origin\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"PGDP_ID\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0msorted_chunks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msorted_chunks\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msorted_origin\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"PGDP_ID\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mmerged_chunk_meta\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted_chunks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0min_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mon\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"PGDP_ID\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/baboondiversity/lib/python3.7/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36mmerge\u001b[0;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m     85\u001b[0m         \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m         \u001b[0mindicator\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindicator\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 87\u001b[0;31m         \u001b[0mvalidate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalidate\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     88\u001b[0m     )\n\u001b[1;32m     89\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/baboondiversity/lib/python3.7/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m    650\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mright_join_keys\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    651\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin_names\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 652\u001b[0;31m         ) = self._get_merge_keys()\n\u001b[0m\u001b[1;32m    653\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m         \u001b[0;31m# validate the merge keys dtypes. We may need to coerce\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/baboondiversity/lib/python3.7/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36m_get_merge_keys\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1003\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mis_rkey\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1004\u001b[0m                         \u001b[0;32mif\u001b[0m \u001b[0mrk\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1005\u001b[0;31m                             \u001b[0mright_keys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mright\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label_or_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1006\u001b[0m                         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1007\u001b[0m                             \u001b[0;31m# work-around for merge_asof(right_index=True)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/baboondiversity/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_get_label_or_level_values\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1556\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1557\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_label_reference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1558\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_check_label_or_level_ambiguity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1559\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mxs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mother_axes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1560\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_level_reference\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/baboondiversity/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_check_label_or_level_ambiguity\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1516\u001b[0m                 \u001b[0;34mf\"{label_article} {label_type} label, which is ambiguous.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1517\u001b[0m             )\n\u001b[0;32m-> 1518\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1519\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1520\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_get_label_or_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mint\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: 'PGDP_ID' is both an index level and a column label, which is ambiguous."
     ]
    }
   ],
   "source": [
    "in_df = meta_data_samples_Sci.loc[meta_data_samples_Sci[\"PGDP_ID\"].isin(chunkcounts.index)]\n",
    "sorted_origin = in_df.sort_values(by=[\"C_origin\"])\n",
    "sorted_chunks = chunkcounts.reindex(sorted_origin[\"PGDP_ID\"])\n",
    "sorted_chunks = sorted_chunks.reindex(columns=sorted_origin[\"PGDP_ID\"])\n",
    "merged_chunk_meta = pd.merge(sorted_chunks, in_df, on=\"PGDP_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=10)\n",
    "X_normalized = preprocessing.normalize(sorted_chunks, norm='l2')\n",
    "projection = pca.fit_transform(X_normalized)\n",
    "\n",
    "print(pca.explained_variance_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcx, pcy = 1, 0\n",
    "g = sns.scatterplot(x=projection[:,pcx], y=projection[:,pcy], hue=merged_chunk_meta[\"C_origin\"])\n",
    "g.legend(bbox_to_anchor=(1.02, 1.02))\n",
    "g.set(xlabel=\"pc{}\".format(pcx+1), ylabel=\"pc{}\".format(pcy+1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I can either generate clusters based on chunkcounts, or try to extract the tree generated by fineSTRUCTURE and analyze that. I will first try to generate it based on chunkcounts/PCA results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkpercentage = chunkcounts.div(chunkcounts.sum(axis=1), axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each individual, I will make a list of individuals over a certain threshold of relatedness."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
