{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this notebook, I will look at the various files generated by the fineSTRUCTURE run on chr8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run ../scripts/notebook_settings.py\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import preprocessing\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['savefig.dpi'] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkcounts_path = \"/home/eriks/baboondiversity/people/eriks/baboon_first_analysis/steps/fs/chr8_no_gelada_linked.chunkcounts.out\"\n",
    "mcmcmatrix_path = \"/home/eriks/baboondiversity/people/eriks/baboon_first_analysis/steps/fs/chr8_no_gelada/chr8_no_gelada_mcmcmatrix.txt\"\n",
    "metadata_path = \"../data/Papio_metadata_with_clustering.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkcounts = pd.read_csv(chunkcounts_path, sep=\" \", header = 1, index_col=0)\n",
    "mcmcmatrix = pd.read_csv(mcmcmatrix_path, sep=\" \", index_col=0)\n",
    "meta_data_samples = pd.read_csv(metadata_path, sep =\" \")\n",
    "meta_data_samples_Sci = meta_data_samples.copy()\n",
    "for i, row in meta_data_samples_Sci.iterrows():\n",
    "    if row.PGDP_ID[0] != \"P\":\n",
    "        meta_data_samples_Sci.at[i, \"PGDP_ID\"] = \"Sci_\"+str(row.PGDP_ID)\n",
    "meta_data_samples_Sci.index = meta_data_samples_Sci.PGDP_ID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determining clusters by considering dendrogram, PCA, fineSTRUCTURE heatmap from chr8 and sampling location."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "kindae = meta_data_samples_Sci.reindex(mcmcmatrix.index)[:27].C_origin\n",
    "cyno_western = meta_data_samples_Sci.reindex(mcmcmatrix.index)[27:37].C_origin\n",
    "ursinus = meta_data_samples_Sci.reindex(mcmcmatrix.index)[37:41].C_origin\n",
    "cyno_ruaha = meta_data_samples_Sci.reindex(mcmcmatrix.index)[41:47].Origin\n",
    "cyno_mikumi = meta_data_samples_Sci.reindex(mcmcmatrix.index)[47:84].Origin \n",
    "cyno_udzungwa = meta_data_samples_Sci.reindex(mcmcmatrix.index)[84:93].Origin #1 Mikumi is also in here PD_0687\n",
    "anubis_manyara = meta_data_samples_Sci.reindex(mcmcmatrix.index)[93:112].Origin\n",
    "anubis_na = meta_data_samples_Sci.reindex(mcmcmatrix.index)[112:122].Origin\n",
    "anubis_tarangire = meta_data_samples_Sci.reindex(mcmcmatrix.index)[122:129].Origin\n",
    "anubis_serengeti = meta_data_samples_Sci.reindex(mcmcmatrix.index)[129:143].Origin\n",
    "anubis_gombe = meta_data_samples_Sci.reindex(mcmcmatrix.index)[143:160].Origin\n",
    "anubis_ethiopia = meta_data_samples_Sci.reindex(mcmcmatrix.index)[160:185].Origin\n",
    "papio_senegal = meta_data_samples_Sci.reindex(mcmcmatrix.index)[185:197].Origin\n",
    "hamadryas_ethiopia = meta_data_samples_Sci.reindex(mcmcmatrix.index)[197:].Origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PD_0730    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0731    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0723    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0735    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0722    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0721    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0725    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0720    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0726    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0727    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0732    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0742    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0729    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0728    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0741    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0743    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0737    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0724    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0733    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0738    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0739    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0740    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0736    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0734    Gog Woreda, Gambella region, Ethiopia\n",
       "PD_0744    Gog Woreda, Gambella region, Ethiopia\n",
       "Name: Origin, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anubis_ethiopia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df = pd.DataFrame()\n",
    "pop_df[\"ID\"] = meta_data_samples_Sci.reindex(mcmcmatrix.index).PGDP_ID\n",
    "pop_df[\"population\"] = \"unknown\"\n",
    "pop_df[\"inclusion\"] = \"unknown\"\n",
    "pop_df.iloc[0:27,1], pop_df.iloc[0:27,2] = \"kindae\", 1\n",
    "pop_df.iloc[27:37,1], pop_df.iloc[27:37,2] = \"cyno_western\", 1\n",
    "pop_df.iloc[37:41,1], pop_df.iloc[37:41,2] = \"ursinus\", 1\n",
    "pop_df.iloc[41:78,1], pop_df.iloc[41:78,2] = \"cyno_mikumi\", 1\n",
    "pop_df.iloc[78:93,1], pop_df.iloc[78:93,2] = \"cyno_RS\", 1\n",
    "pop_df.iloc[93:110,1], pop_df.iloc[93:110,2] = \"anubis_gombe\", 1\n",
    "pop_df.iloc[110:112,1], pop_df.iloc[110:112,2] = \"anubis_kenya\", 0\n",
    "pop_df.iloc[112:126,1], pop_df.iloc[112:126,2] = \"anubis_serengeti\", 1\n",
    "pop_df.iloc[126:162,1], pop_df.iloc[126:162,2] = \"anubis_MGAT\", 1\n",
    "pop_df.iloc[162:187,1], pop_df.iloc[162:187,2] = \"anubis_ethiopia\", 1\n",
    "pop_df.iloc[187:189,1], pop_df.iloc[187:189,2] = \"gelada\", 0\n",
    "pop_df.iloc[189:201,1], pop_df.iloc[189:201,2] = \"papio\", 1\n",
    "pop_df.iloc[201:,1], pop_df.iloc[201:,2] = \"hamadryas\", 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zero_PGDP_ID = []\n",
    "for population in pop_df.population.unique():\n",
    "    sub_df = pop_df.loc[pop_df.population==population]\n",
    "    zero_PGDP_ID.extend(sub_df[\"ID\"].sample(n=max(len(sub_df), 4)-4))\n",
    "pop_df_subset = pop_df.copy()\n",
    "for ID in zero_PGDP_ID:\n",
    "    pop_df_subset.loc[pop_df_subset.ID==ID, \"inclusion\"] = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_reorder = pop_df.reindex(meta_data_samples_Sci.index)\n",
    "pop_df_reorder.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/idfile_fs_cluster.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_subset_reorder = pop_df_subset.reindex(meta_data_samples_Sci.index)\n",
    "pop_df_subset_reorder.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/idfile_subset.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Subset for only the females in the chrX vcf. Alan has followed my gendering, so that is nice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_x = \"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/chrX/chrX.phased.rehead.vcf.gz\"\n",
    "callset = allel.read_vcf(vcf_x)\n",
    "pop_df_females = pop_df.reindex(callset[\"samples\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_females.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/chrX_females.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_females"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Idfile containing only the 8 clusters originally determined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df = pd.DataFrame()\n",
    "pop_df[\"ID\"] = meta_data_samples_Sci.reindex(mcmcmatrix.index).PGDP_ID\n",
    "pop_df[\"population\"] = \"unknown\"\n",
    "pop_df[\"inclusion\"] = \"unknown\"\n",
    "pop_df.iloc[0:27,1], pop_df.iloc[0:27,2] = \"kindae\", 1\n",
    "pop_df.iloc[27:37,1], pop_df.iloc[27:37,2] = \"cyno_western\", 1\n",
    "pop_df.iloc[37:41,1], pop_df.iloc[37:41,2] = \"ursinus\", 1\n",
    "pop_df.iloc[41:93,1], pop_df.iloc[41:93,2] = \"cyno_central\", 1\n",
    "pop_df.iloc[93:110,1], pop_df.iloc[93:110,2] = \"anubis_tanzania\", 1\n",
    "pop_df.iloc[110:112,1], pop_df.iloc[110:112,2] = \"anubis_kenya\", 0\n",
    "pop_df.iloc[112:126,1], pop_df.iloc[112:126,2] = \"anubis_tanzania\", 1\n",
    "pop_df.iloc[126:162,1], pop_df.iloc[126:162,2] = \"anubis_tanzania\", 1\n",
    "pop_df.iloc[162:187,1], pop_df.iloc[162:187,2] = \"anubis_ethiopia\", 1\n",
    "pop_df.iloc[187:189,1], pop_df.iloc[187:189,2] = \"gelada\", 0\n",
    "pop_df.iloc[189:201,1], pop_df.iloc[189:201,2] = \"papio\", 1\n",
    "pop_df.iloc[201:,1], pop_df.iloc[201:,2] = \"hamadryas\", 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_reorder = pop_df.reindex(meta_data_samples_Sci.index)\n",
    "pop_df_reorder.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/idfile_8_cluster.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "zero_PGDP_ID = []\n",
    "for population in pop_df.population.unique():\n",
    "    sub_df = pop_df.loc[pop_df.population==population]\n",
    "    zero_PGDP_ID.extend(sub_df[\"ID\"].sample(n=max(len(sub_df), 10)-10))\n",
    "pop_df_subset = pop_df.copy()\n",
    "for ID in zero_PGDP_ID:\n",
    "    pop_df_subset.loc[pop_df_subset.ID==ID, \"inclusion\"] = 0\n",
    "pop_df_subset_reorder = pop_df_subset.reindex(meta_data_samples_Sci.index)\n",
    "pop_df_subset_reorder.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/idfile_8_cluster_subset.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Repeat for 8 cluster, females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_x = \"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/chrX/chrX.phased.rehead.vcf.gz\"\n",
    "callset = allel.read_vcf(vcf_x)\n",
    "pop_df_females = pop_df.reindex(callset[\"samples\"])\n",
    "pop_df_females.to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/chrX_8_cluster_females.ids\",\n",
    "              sep=\" \", header=False, index=False)\n",
    "\n",
    "pop_df_females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/bcf_inclusion.txt\", \"w\")\n",
    "included_inds = (pop_df_females.loc[pop_df_females.inclusion==1].ID.values)\n",
    "for ind in included_inds:\n",
    "    f.write(ind+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_df_females.loc[pop_df_females.inclusion==1].to_csv(\"/home/eriks/baboondiversity/data/PG_panu3_phased_chromosomes_4_7_2021/included_females.ids\",\n",
    "              sep=\" \", header=False, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\",\".join(pop_df_females.loc[pop_df_females.inclusion==1].ID.values)+\"\\n\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkcounts.loc[chunkcounts.max() > 4000]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_df = meta_data_samples_Sci.loc[meta_data_samples_Sci[\"PGDP_ID\"].isin(chunkcounts.index)]\n",
    "sorted_origin = in_df.sort_values(by=[\"C_origin\"])\n",
    "sorted_chunks = chunkcounts.reindex(sorted_origin[\"PGDP_ID\"])\n",
    "sorted_chunks = sorted_chunks.reindex(columns=sorted_origin[\"PGDP_ID\"])\n",
    "merged_chunk_meta = pd.merge(sorted_chunks, in_df, on=\"PGDP_ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=10)\n",
    "X_normalized = preprocessing.normalize(sorted_chunks, norm='l2')\n",
    "projection = pca.fit_transform(X_normalized)\n",
    "\n",
    "print(pca.explained_variance_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcx, pcy = 1, 0\n",
    "g = sns.scatterplot(x=projection[:,pcx], y=projection[:,pcy], hue=merged_chunk_meta[\"C_origin\"])\n",
    "g.legend(bbox_to_anchor=(1.02, 1.02))\n",
    "g.set(xlabel=\"pc{}\".format(pcx+1), ylabel=\"pc{}\".format(pcy+1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I can either generate clusters based on chunkcounts, or try to extract the tree generated by fineSTRUCTURE and analyze that. I will first try to generate it based on chunkcounts/PCA results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "chunkpercentage = chunkcounts.div(chunkcounts.sum(axis=1), axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each individual, I will make a list of individuals over a certain threshold of relatedness."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
