{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Initial configuration, probably overkill in imports.\n",
    "import sys, os, re\n",
    "from functools import reduce\n",
    "import glob\n",
    "import numpy as np\n",
    "import allel\n",
    "import zarr\n",
    "import dask\n",
    "import numcodecs\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "from horizonplot import horizonplot\n",
    "\n",
    "#os.environ[\"MODIN_ENGINE\"] = \"ray\"\n",
    "\n",
    "#import modin.pandas as pd\n",
    "import pandas as pd\n",
    "\n",
    "%matplotlib inline\n",
    "from IPython.display import set_matplotlib_formats\n",
    "set_matplotlib_formats('retina', 'png')\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.colors import ListedColormap\n",
    "import seaborn as sns\n",
    "sns.set()\n",
    "sns.set_theme()\n",
    "sns.set_style(\"white\")\n",
    "sns.set_context(\"notebook\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_dir = \"/faststorage/project/primatediversity/data/PGDP_16_7_2020/variants/\"\n",
    "vcf_suffix = \".variable.filtered.HF.snps.vcf.gz\"\n",
    "metadata = \"/faststorage/project/primatediversity/data/PGDP_16_7_2020/metadata/PGDP_md.16.7.2020.xlsx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_vs_A = glob.glob(\"/faststorage/project/primatediversity/data/PGDP_16_7_2020/metadata/*simplified.csv\")\n",
    "sex_via_cov = \"/faststorage/project/primatediversity/data/PGDP_16_7_2020/metadata/sex_via_coverage_Lukas.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PGDP_ID</th>\n",
       "      <th>species</th>\n",
       "      <th>autosomal_coverage</th>\n",
       "      <th>X_coverage</th>\n",
       "      <th>X_over_autosome_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_0001</td>\n",
       "      <td>Pithecia_pithecia.fasta</td>\n",
       "      <td>31</td>\n",
       "      <td>30</td>\n",
       "      <td>0.967742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_0002</td>\n",
       "      <td>Callithrix_jacchus.fasta</td>\n",
       "      <td>24</td>\n",
       "      <td>12</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_0003</td>\n",
       "      <td>Pithecia_pithecia.fasta</td>\n",
       "      <td>30</td>\n",
       "      <td>16</td>\n",
       "      <td>0.533333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_0004</td>\n",
       "      <td>Atele_fusciceps.fasta</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_0005</td>\n",
       "      <td>Atele_fusciceps.fasta</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>673</th>\n",
       "      <td>SAMN01920547</td>\n",
       "      <td>Pongo_pygmaeus.fasta</td>\n",
       "      <td>25</td>\n",
       "      <td>26</td>\n",
       "      <td>1.040000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674</th>\n",
       "      <td>SAMN01920548</td>\n",
       "      <td>Pongo_pygmaeus.fasta</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675</th>\n",
       "      <td>SAMN01920549</td>\n",
       "      <td>Pongo_pygmaeus.fasta</td>\n",
       "      <td>30</td>\n",
       "      <td>31</td>\n",
       "      <td>1.033330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>SAMN01920550</td>\n",
       "      <td>Pongo_pygmaeus.fasta</td>\n",
       "      <td>29</td>\n",
       "      <td>30</td>\n",
       "      <td>1.034480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>SAMN01920551</td>\n",
       "      <td>Pongo_pygmaeus.fasta</td>\n",
       "      <td>29</td>\n",
       "      <td>15</td>\n",
       "      <td>0.517241</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>678 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          PGDP_ID                   species  autosomal_coverage  X_coverage  \\\n",
       "0         PD_0001   Pithecia_pithecia.fasta                  31          30   \n",
       "1         PD_0002  Callithrix_jacchus.fasta                  24          12   \n",
       "2         PD_0003   Pithecia_pithecia.fasta                  30          16   \n",
       "3         PD_0004     Atele_fusciceps.fasta                  31          31   \n",
       "4         PD_0005     Atele_fusciceps.fasta                  30          15   \n",
       "..            ...                       ...                 ...         ...   \n",
       "673  SAMN01920547      Pongo_pygmaeus.fasta                  25          26   \n",
       "674  SAMN01920548      Pongo_pygmaeus.fasta                  26          26   \n",
       "675  SAMN01920549      Pongo_pygmaeus.fasta                  30          31   \n",
       "676  SAMN01920550      Pongo_pygmaeus.fasta                  29          30   \n",
       "677  SAMN01920551      Pongo_pygmaeus.fasta                  29          15   \n",
       "\n",
       "     X_over_autosome_ratio  \n",
       "0                 0.967742  \n",
       "1                 0.500000  \n",
       "2                 0.533333  \n",
       "3                 1.000000  \n",
       "4                 0.500000  \n",
       "..                     ...  \n",
       "673               1.040000  \n",
       "674               1.000000  \n",
       "675               1.033330  \n",
       "676               1.034480  \n",
       "677               0.517241  \n",
       "\n",
       "[678 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d_cov = pd.read_csv(sex_via_cov)\n",
    "d_cov[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Atele_fusciceps.fasta', 'Aotus_nancymaae.fasta',\n",
       "       'Pithecia_pithecia.fasta', 'Cebus_albifrons.fasta',\n",
       "       'Cercopithecus_mitis.fasta', 'Gorilla_gorilla_gorilla.fasta',\n",
       "       'Hoolock_hoolock.fasta', 'Macaca_mulatta.fasta',\n",
       "       'Nycticebus_pygmaeus.fasta', 'Pan_troglodytes.fasta',\n",
       "       'rheMac10.fa', nan, 'Papio_anubis.fasta',\n",
       "       'Pygathrix_nemaeus.fasta', 'Sapajus_apella.fasta',\n",
       "       'Rhinopithecus_roxellana.fasta', 'Trachypithecus_phayrei.fasta'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hetero = \"/faststorage/project/primatediversity/data/PGDP_16_7_2020/metadata/sex_Marjo_heterozygosity.csv\"\n",
    "d_hetero = pd.read_csv(hetero, \"\\t\")\n",
    "d_hetero.reference_file.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A small test to understand merge, fillna and sort_values before I run it on the real dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1={\"pos\": [1, 10], \"ID_1\": [0, 1]}\n",
    "d2={\"pos\": [1, 9], \"ID_2\": [0, 1]}\n",
    "df1 = pd.DataFrame(data=d1)\n",
    "df2 = pd.DataFrame(data=d2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ID_1</th>\n",
       "      <th>ID_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos  ID_1  ID_2\n",
       "0    1   0.0   0.0\n",
       "2    9   0.0   1.0\n",
       "1   10   1.0   0.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_df = pd.merge(df1, df2, on=\"pos\", how=\"outer\")\n",
    "merge_df.fillna(0, inplace=True)\n",
    "merge_df.sort_values(\"pos\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code extracts the het sites and their positions. As the sites might vary among individuals, there is a pos-based merge to generate the final dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Atele_fusciceps.fasta\n",
      "Aotus_nancymaae.fasta\n",
      "Pithecia_pithecia.fasta\n",
      "tarseq_214 PD_0001\n",
      "tarseq_56 PD_0001\n",
      "tarseq_87 PD_0001\n",
      "Cebus_albifrons.fasta\n",
      "Contig110 PD_0368\n",
      "Cercopithecus_mitis.fasta\n",
      "Gorilla_gorilla_gorilla.fasta\n",
      "Hoolock_hoolock.fasta\n",
      "Macaca_mulatta.fasta\n",
      "Nycticebus_pygmaeus.fasta\n",
      "Pan_troglodytes.fasta\n",
      "rheMac10.fa\n",
      "nan\n",
      "Papio_anubis.fasta\n",
      "Pygathrix_nemaeus.fasta\n",
      "Sapajus_apella.fasta\n",
      "Rhinopithecus_roxellana.fasta\n",
      "Trachypithecus_phayrei.fasta\n"
     ]
    }
   ],
   "source": [
    "simplified_csv = \"/faststorage/project/primatediversity/data/PGDP_16_7_2020/metadata/{}_X_vs_A_simplified.csv\"\n",
    "\n",
    "def count_het_species(d, genus):\n",
    "    X_vs_A = simplified_csv.format(genus[:-6])\n",
    "    contigs = pd.read_csv(X_vs_A, sep=\"\\t\")\n",
    "    x_contigs = contigs.loc[contigs.decision == \"X\"].scaffold.values\n",
    "    df_list = []\n",
    "    for contig in x_contigs:\n",
    "        df_list_contig = []\n",
    "        for i, row in d.iterrows():\n",
    "            ID = row.PDGP_ID\n",
    "            path = vcf_dir+ID+vcf_suffix\n",
    "            if os.path.exists(path):\n",
    "                vcf_callset = allel.read_vcf(path, region=contig, fields=[\"calldata/GT\", \"variants/POS\"])\n",
    "                if vcf_callset != None:\n",
    "                    gt = allel.GenotypeArray(vcf_callset[\"calldata/GT\"])\n",
    "                    pos = vcf_callset[\"variants/POS\"]\n",
    "                    het_state = gt.is_het()\n",
    "                    dfgt = pd.DataFrame(het_state, columns=[\"{}\".format(ID)])\n",
    "                    dfpos = pd.DataFrame(pos, columns=[\"pos\"])\n",
    "                    dfpos = dfpos.loc[dfgt[ID] == True]\n",
    "                    dfgt = dfgt.loc[dfgt[ID] == True]\n",
    "                    df_list_contig.append(pd.concat([dfpos, dfgt], axis=1))\n",
    "                else:\n",
    "                    print(contig, ID)\n",
    "            else:\n",
    "                  print(ID)\n",
    "        #Merging these dataframes, with differing sizes. Merging here so it can handle multiple contigs\n",
    "        df_list.append(reduce(lambda x, y: pd.merge(x, y, on = 'pos', how=\"outer\"), df_list_contig))\n",
    "    df = pd.concat(df_list, ignore_index=True)\n",
    "    df.reset_index()\n",
    "    return df.fillna(False)\n",
    "\n",
    "genus_dict = {}\n",
    "for genus in d_hetero.reference_file.unique():\n",
    "    d = d_hetero.loc[d_hetero.reference_file == genus]\n",
    "    print(genus)\n",
    "    if genus not in [\"rheMac10.fa\", \"Papio_anubis.fasta\"] and len(d) > 0:\n",
    "        df = count_het_species(d, genus)\n",
    "        if len(df) > 0:\n",
    "            genus_dict[genus] = df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, I will split the dataframe based on Marjorlaines sexing, and generate a M+1*F+1 size dataframe, with M being number of males and F being number of females, and count the nunber of sites fitting into the various classes.\n",
    "\n",
    "I choose to subsample so that there is an equal number of males/females. This removes information, but makes it much more comparable when looking at males/females."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 11\n"
     ]
    }
   ],
   "source": [
    "np.set_printoptions(suppress=True)\n",
    "\n",
    "def generate_shared_het_matrix(d_hetero, df):\n",
    "    d_m = d.loc[d.sex_coverage == \"male\"]\n",
    "    d_f = d.loc[d.sex_coverage == \"female\"]\n",
    "    min_s = min(len(d_m), len(d_f))\n",
    "    m_l = d_m.PDGP_ID.tolist()\n",
    "    f_l = d_f.PDGP_ID.tolist()\n",
    "    m_df = df.loc[:, m_l[:min_s]]\n",
    "    f_df = df.loc[:, f_l[:min_s]]\n",
    "    #print(m_df.sum(axis=1))\n",
    "    #print(f_df.sum(axis=1))\n",
    "    df.reset_index() #To handle an error I probably got due to the earlier concats.\n",
    "    print(len(d_m), len(d_f))\n",
    "    matrix = np.zeros((min_s+1, min_s+1))\n",
    "    m_sum = m_df.sum(axis=1)\n",
    "    f_sum = f_df.sum(axis=1)\n",
    "    for i in range(len(df)):\n",
    "        matrix[m_sum[i], f_sum[i]] += 1\n",
    "    matrix[0, 0] = 0\n",
    "    return matrix\n",
    "\n",
    "def summarize_matrix(matrix, genus):\n",
    "    singleton_male, singleton_female = matrix.sum(axis=1)[1], matrix.sum(axis=1)[2]\n",
    "    d = {\"genus\": genus}\n",
    "\n",
    "d = d_hetero.loc[(d_hetero.PDGP_ID != \"PD_0001\") & (d_hetero.PDGP_ID != \"PD_0368\")]\n",
    "d = d.loc[d.reference_file == \"Gorilla_gorilla_gorilla.fasta\"]\n",
    "n_matrix = generate_shared_het_matrix(d, genus_dict[\"Gorilla_gorilla_gorilla.fasta\"])\n",
    "#summarize_matrix(n_matrix, \"Gorilla_gorilla_gorilla.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5 11\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genus</th>\n",
       "      <th>n</th>\n",
       "      <th>sex</th>\n",
       "      <th>s_het</th>\n",
       "      <th>average_het</th>\n",
       "      <th>singletons</th>\n",
       "      <th>all_shared_het</th>\n",
       "      <th>above_hw</th>\n",
       "      <th>weight_above_hw</th>\n",
       "      <th>total_het</th>\n",
       "      <th>both_f_m</th>\n",
       "      <th>het_all</th>\n",
       "      <th>above_hw_m_f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>5</td>\n",
       "      <td>male</td>\n",
       "      <td>178461.0</td>\n",
       "      <td>114810.6</td>\n",
       "      <td>0.246844</td>\n",
       "      <td>0.352794</td>\n",
       "      <td>0.496590</td>\n",
       "      <td>0.727194</td>\n",
       "      <td>551534.0</td>\n",
       "      <td>0.032045</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>5</td>\n",
       "      <td>female</td>\n",
       "      <td>390747.0</td>\n",
       "      <td>153233.6</td>\n",
       "      <td>0.411770</td>\n",
       "      <td>0.013280</td>\n",
       "      <td>0.086731</td>\n",
       "      <td>0.183705</td>\n",
       "      <td>551534.0</td>\n",
       "      <td>0.032045</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     genus  n     sex     s_het  average_het  singletons  \\\n",
       "0  Gorilla_gorilla_gorilla  5    male  178461.0     114810.6    0.246844   \n",
       "1  Gorilla_gorilla_gorilla  5  female  390747.0     153233.6    0.411770   \n",
       "\n",
       "   all_shared_het  above_hw  weight_above_hw  total_het  both_f_m   het_all  \\\n",
       "0        0.352794  0.496590         0.727194   551534.0  0.032045  0.000225   \n",
       "1        0.013280  0.086731         0.183705   551534.0  0.032045  0.000225   \n",
       "\n",
       "   above_hw_m_f  \n",
       "0      0.005882  \n",
       "1      0.005882  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def summarize_matrix(matrix, genus):\n",
    "    df_list = []\n",
    "    for g, a in [[\"male\", 1], [\"female\", 0]]:\n",
    "        i = len(matrix)-1\n",
    "        total_het = matrix.sum()\n",
    "        singleton = matrix.sum(axis=a)[1]\n",
    "        s_het = sum(matrix.sum(axis=a)[1:])\n",
    "        all_shared = matrix.sum(axis=a)[i]\n",
    "        above_hw = sum(matrix.sum(axis=a)[1+i//2+(i%2>0):])\n",
    "        het_per_individual = sum(matrix.sum(axis=a)[1:]*list(range(1, i+1)))\n",
    "        calls_in_above_hw = sum(matrix.sum(axis=a)[1+i//2+(i%2>0):]*list(range(1+i//2+(i%2>0), i+1)))\n",
    "        both_f_m = matrix[1:, 1:].sum()\n",
    "        all_f_m = matrix[i:, i:].sum()\n",
    "        above_hw_m_f = matrix[1+i//2+(i%2>0):, 1:].sum()\n",
    "        d = {\"genus\": genus[:-6], \"n\": i, \"sex\": g, \"s_het\": s_het, \"average_het\": het_per_individual/i,\n",
    "             \"singletons\": singleton/s_het, \"all_shared_het\": all_shared/s_het, \"above_hw\": above_hw/s_het, \n",
    "             \"weight_above_hw\": calls_in_above_hw/het_per_individual, \n",
    "             \"total_het\": total_het,\"both_f_m\": both_f_m/total_het, \"het_all\": all_f_m/total_het,\n",
    "             \"above_hw_m_f\": above_hw_m_f/total_het}\n",
    "        df = pd.DataFrame(data=d, index=[0])\n",
    "        df_list.append(df)\n",
    "    df = pd.concat(df_list, ignore_index=True)\n",
    "    return df\n",
    "d = d_hetero.loc[(d_hetero.PDGP_ID != \"PD_0001\") & (d_hetero.PDGP_ID != \"PD_0368\")]\n",
    "d = d.loc[d.reference_file == \"Gorilla_gorilla_gorilla.fasta\"]\n",
    "n_matrix = generate_shared_het_matrix(d, genus_dict[\"Gorilla_gorilla_gorilla.fasta\"])\n",
    "summarize_matrix(n_matrix, \"Gorilla_gorilla_gorilla.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[404., 577., 496., 268.,  73.],\n",
       "       [382., 363., 361., 196., 124.]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n_matrix[4:, 1:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Atele_fusciceps.fasta\n",
      "16 14\n",
      "Aotus_nancymaae.fasta\n",
      "17 6\n",
      "Pithecia_pithecia.fasta\n",
      "21 23\n",
      "Cebus_albifrons.fasta\n",
      "12 7\n",
      "Cercopithecus_mitis.fasta\n",
      "1 3\n",
      "Gorilla_gorilla_gorilla.fasta\n",
      "5 11\n",
      "Hoolock_hoolock.fasta\n",
      "3 6\n",
      "Macaca_mulatta.fasta\n",
      "35 29\n",
      "Nycticebus_pygmaeus.fasta\n",
      "1 3\n",
      "Pan_troglodytes.fasta\n",
      "2 2\n",
      "Pygathrix_nemaeus.fasta\n",
      "7 3\n",
      "Sapajus_apella.fasta\n",
      "11 3\n",
      "Rhinopithecus_roxellana.fasta\n",
      "6 4\n",
      "Trachypithecus_phayrei.fasta\n",
      "5 1\n"
     ]
    }
   ],
   "source": [
    "df_list = []\n",
    "for genus in genus_dict:\n",
    "    print(genus)\n",
    "    d = d_hetero.loc[(d_hetero.PDGP_ID != \"PD_0001\") & (d_hetero.PDGP_ID != \"PD_0368\")]\n",
    "    d = d.loc[d.reference_file == genus]\n",
    "    n_matrix = generate_shared_het_matrix(d, genus_dict[genus])\n",
    "    df = summarize_matrix(n_matrix, genus)\n",
    "    df_list.append(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genus</th>\n",
       "      <th>n</th>\n",
       "      <th>sex</th>\n",
       "      <th>s_het</th>\n",
       "      <th>average_het</th>\n",
       "      <th>singletons</th>\n",
       "      <th>all_shared_het</th>\n",
       "      <th>above_hw</th>\n",
       "      <th>weight_above_hw</th>\n",
       "      <th>total_het</th>\n",
       "      <th>both_f_m</th>\n",
       "      <th>het_all</th>\n",
       "      <th>above_hw_m_f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Atele_fusciceps</td>\n",
       "      <td>14</td>\n",
       "      <td>male</td>\n",
       "      <td>77537</td>\n",
       "      <td>13911</td>\n",
       "      <td>0.566916</td>\n",
       "      <td>0.003908</td>\n",
       "      <td>0.080349</td>\n",
       "      <td>0.321127</td>\n",
       "      <td>399595</td>\n",
       "      <td>0.091703</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.014802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Atele_fusciceps</td>\n",
       "      <td>14</td>\n",
       "      <td>female</td>\n",
       "      <td>358702</td>\n",
       "      <td>44904</td>\n",
       "      <td>0.603473</td>\n",
       "      <td>0.000050</td>\n",
       "      <td>0.008458</td>\n",
       "      <td>0.044298</td>\n",
       "      <td>399595</td>\n",
       "      <td>0.091703</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.014802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aotus_nancymaae</td>\n",
       "      <td>6</td>\n",
       "      <td>male</td>\n",
       "      <td>24786</td>\n",
       "      <td>10411</td>\n",
       "      <td>0.477447</td>\n",
       "      <td>0.126200</td>\n",
       "      <td>0.284394</td>\n",
       "      <td>0.587242</td>\n",
       "      <td>105367</td>\n",
       "      <td>0.084761</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.035172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Aotus_nancymaae</td>\n",
       "      <td>6</td>\n",
       "      <td>female</td>\n",
       "      <td>89512</td>\n",
       "      <td>24792</td>\n",
       "      <td>0.672837</td>\n",
       "      <td>0.009317</td>\n",
       "      <td>0.103327</td>\n",
       "      <td>0.271532</td>\n",
       "      <td>105367</td>\n",
       "      <td>0.084761</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.035172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pithecia_pithecia</td>\n",
       "      <td>21</td>\n",
       "      <td>male</td>\n",
       "      <td>127094</td>\n",
       "      <td>15714</td>\n",
       "      <td>0.498765</td>\n",
       "      <td>0.000889</td>\n",
       "      <td>0.016185</td>\n",
       "      <td>0.091210</td>\n",
       "      <td>686421</td>\n",
       "      <td>0.087024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pithecia_pithecia</td>\n",
       "      <td>21</td>\n",
       "      <td>female</td>\n",
       "      <td>619062</td>\n",
       "      <td>45311</td>\n",
       "      <td>0.675259</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.001205</td>\n",
       "      <td>0.010409</td>\n",
       "      <td>686421</td>\n",
       "      <td>0.087024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cebus_albifrons</td>\n",
       "      <td>7</td>\n",
       "      <td>male</td>\n",
       "      <td>18654</td>\n",
       "      <td>4522</td>\n",
       "      <td>0.684089</td>\n",
       "      <td>0.002359</td>\n",
       "      <td>0.065884</td>\n",
       "      <td>0.217316</td>\n",
       "      <td>60073</td>\n",
       "      <td>0.102375</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.017063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Cebus_albifrons</td>\n",
       "      <td>7</td>\n",
       "      <td>female</td>\n",
       "      <td>47569</td>\n",
       "      <td>9177</td>\n",
       "      <td>0.727217</td>\n",
       "      <td>0.000799</td>\n",
       "      <td>0.004225</td>\n",
       "      <td>0.017854</td>\n",
       "      <td>60073</td>\n",
       "      <td>0.102375</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.017063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Cercopithecus_mitis</td>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>21040</td>\n",
       "      <td>21040</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>143761</td>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Cercopithecus_mitis</td>\n",
       "      <td>1</td>\n",
       "      <td>female</td>\n",
       "      <td>124991</td>\n",
       "      <td>124991</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>143761</td>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>5</td>\n",
       "      <td>male</td>\n",
       "      <td>178461</td>\n",
       "      <td>114810</td>\n",
       "      <td>0.246844</td>\n",
       "      <td>0.352794</td>\n",
       "      <td>0.496590</td>\n",
       "      <td>0.727194</td>\n",
       "      <td>551534</td>\n",
       "      <td>0.032045</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Gorilla_gorilla_gorilla</td>\n",
       "      <td>5</td>\n",
       "      <td>female</td>\n",
       "      <td>390747</td>\n",
       "      <td>153233</td>\n",
       "      <td>0.411770</td>\n",
       "      <td>0.013280</td>\n",
       "      <td>0.086731</td>\n",
       "      <td>0.183705</td>\n",
       "      <td>551534</td>\n",
       "      <td>0.032045</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Hoolock_hoolock</td>\n",
       "      <td>3</td>\n",
       "      <td>male</td>\n",
       "      <td>6034</td>\n",
       "      <td>3733</td>\n",
       "      <td>0.456414</td>\n",
       "      <td>0.312562</td>\n",
       "      <td>0.312562</td>\n",
       "      <td>0.505179</td>\n",
       "      <td>27102</td>\n",
       "      <td>0.133496</td>\n",
       "      <td>0.019777</td>\n",
       "      <td>0.059848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Hoolock_hoolock</td>\n",
       "      <td>3</td>\n",
       "      <td>female</td>\n",
       "      <td>24686</td>\n",
       "      <td>10467</td>\n",
       "      <td>0.766629</td>\n",
       "      <td>0.038726</td>\n",
       "      <td>0.038726</td>\n",
       "      <td>0.091329</td>\n",
       "      <td>27102</td>\n",
       "      <td>0.133496</td>\n",
       "      <td>0.019777</td>\n",
       "      <td>0.059848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Macaca_mulatta</td>\n",
       "      <td>29</td>\n",
       "      <td>male</td>\n",
       "      <td>176737</td>\n",
       "      <td>23372</td>\n",
       "      <td>0.466597</td>\n",
       "      <td>0.001539</td>\n",
       "      <td>0.054680</td>\n",
       "      <td>0.299281</td>\n",
       "      <td>1916367</td>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Macaca_mulatta</td>\n",
       "      <td>29</td>\n",
       "      <td>female</td>\n",
       "      <td>1819955</td>\n",
       "      <td>121955</td>\n",
       "      <td>0.526172</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.002071</td>\n",
       "      <td>0.021359</td>\n",
       "      <td>1916367</td>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Nycticebus_pygmaeus</td>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>76282</td>\n",
       "      <td>76282</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>180589</td>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Nycticebus_pygmaeus</td>\n",
       "      <td>1</td>\n",
       "      <td>female</td>\n",
       "      <td>129768</td>\n",
       "      <td>129768</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>180589</td>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Pan_troglodytes</td>\n",
       "      <td>2</td>\n",
       "      <td>male</td>\n",
       "      <td>17047</td>\n",
       "      <td>11423</td>\n",
       "      <td>0.659764</td>\n",
       "      <td>0.340236</td>\n",
       "      <td>0.340236</td>\n",
       "      <td>0.507725</td>\n",
       "      <td>250509</td>\n",
       "      <td>0.022893</td>\n",
       "      <td>0.003249</td>\n",
       "      <td>0.010071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Pan_troglodytes</td>\n",
       "      <td>2</td>\n",
       "      <td>female</td>\n",
       "      <td>239197</td>\n",
       "      <td>123067</td>\n",
       "      <td>0.970995</td>\n",
       "      <td>0.029005</td>\n",
       "      <td>0.029005</td>\n",
       "      <td>0.056376</td>\n",
       "      <td>250509</td>\n",
       "      <td>0.022893</td>\n",
       "      <td>0.003249</td>\n",
       "      <td>0.010071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Pygathrix_nemaeus</td>\n",
       "      <td>3</td>\n",
       "      <td>male</td>\n",
       "      <td>2277</td>\n",
       "      <td>1290</td>\n",
       "      <td>0.547211</td>\n",
       "      <td>0.247255</td>\n",
       "      <td>0.247255</td>\n",
       "      <td>0.436321</td>\n",
       "      <td>21480</td>\n",
       "      <td>0.050047</td>\n",
       "      <td>0.002793</td>\n",
       "      <td>0.017691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pygathrix_nemaeus</td>\n",
       "      <td>3</td>\n",
       "      <td>female</td>\n",
       "      <td>20278</td>\n",
       "      <td>8559</td>\n",
       "      <td>0.761663</td>\n",
       "      <td>0.028011</td>\n",
       "      <td>0.028011</td>\n",
       "      <td>0.066358</td>\n",
       "      <td>21480</td>\n",
       "      <td>0.050047</td>\n",
       "      <td>0.002793</td>\n",
       "      <td>0.017691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Sapajus_apella</td>\n",
       "      <td>3</td>\n",
       "      <td>male</td>\n",
       "      <td>816</td>\n",
       "      <td>411</td>\n",
       "      <td>0.642157</td>\n",
       "      <td>0.154412</td>\n",
       "      <td>0.154412</td>\n",
       "      <td>0.306321</td>\n",
       "      <td>4607</td>\n",
       "      <td>0.067072</td>\n",
       "      <td>0.003690</td>\n",
       "      <td>0.020404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Sapajus_apella</td>\n",
       "      <td>3</td>\n",
       "      <td>female</td>\n",
       "      <td>4100</td>\n",
       "      <td>1575</td>\n",
       "      <td>0.856829</td>\n",
       "      <td>0.009268</td>\n",
       "      <td>0.009268</td>\n",
       "      <td>0.024127</td>\n",
       "      <td>4607</td>\n",
       "      <td>0.067072</td>\n",
       "      <td>0.003690</td>\n",
       "      <td>0.020404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Rhinopithecus_roxellana</td>\n",
       "      <td>4</td>\n",
       "      <td>male</td>\n",
       "      <td>54115</td>\n",
       "      <td>29342</td>\n",
       "      <td>0.417186</td>\n",
       "      <td>0.217555</td>\n",
       "      <td>0.368548</td>\n",
       "      <td>0.610074</td>\n",
       "      <td>319596</td>\n",
       "      <td>0.050360</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.022488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Rhinopithecus_roxellana</td>\n",
       "      <td>4</td>\n",
       "      <td>female</td>\n",
       "      <td>281576</td>\n",
       "      <td>114378</td>\n",
       "      <td>0.537187</td>\n",
       "      <td>0.001868</td>\n",
       "      <td>0.160149</td>\n",
       "      <td>0.296840</td>\n",
       "      <td>319596</td>\n",
       "      <td>0.050360</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.022488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Trachypithecus_phayrei</td>\n",
       "      <td>1</td>\n",
       "      <td>male</td>\n",
       "      <td>10339</td>\n",
       "      <td>10339</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26935</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Trachypithecus_phayrei</td>\n",
       "      <td>1</td>\n",
       "      <td>female</td>\n",
       "      <td>20475</td>\n",
       "      <td>20475</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>26935</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      genus   n     sex    s_het  average_het  singletons  \\\n",
       "0           Atele_fusciceps  14    male    77537        13911    0.566916   \n",
       "1           Atele_fusciceps  14  female   358702        44904    0.603473   \n",
       "2           Aotus_nancymaae   6    male    24786        10411    0.477447   \n",
       "3           Aotus_nancymaae   6  female    89512        24792    0.672837   \n",
       "4         Pithecia_pithecia  21    male   127094        15714    0.498765   \n",
       "5         Pithecia_pithecia  21  female   619062        45311    0.675259   \n",
       "6           Cebus_albifrons   7    male    18654         4522    0.684089   \n",
       "7           Cebus_albifrons   7  female    47569         9177    0.727217   \n",
       "8       Cercopithecus_mitis   1    male    21040        21040    1.000000   \n",
       "9       Cercopithecus_mitis   1  female   124991       124991    1.000000   \n",
       "10  Gorilla_gorilla_gorilla   5    male   178461       114810    0.246844   \n",
       "11  Gorilla_gorilla_gorilla   5  female   390747       153233    0.411770   \n",
       "12          Hoolock_hoolock   3    male     6034         3733    0.456414   \n",
       "13          Hoolock_hoolock   3  female    24686        10467    0.766629   \n",
       "14           Macaca_mulatta  29    male   176737        23372    0.466597   \n",
       "15           Macaca_mulatta  29  female  1819955       121955    0.526172   \n",
       "16      Nycticebus_pygmaeus   1    male    76282        76282    1.000000   \n",
       "17      Nycticebus_pygmaeus   1  female   129768       129768    1.000000   \n",
       "18          Pan_troglodytes   2    male    17047        11423    0.659764   \n",
       "19          Pan_troglodytes   2  female   239197       123067    0.970995   \n",
       "20        Pygathrix_nemaeus   3    male     2277         1290    0.547211   \n",
       "21        Pygathrix_nemaeus   3  female    20278         8559    0.761663   \n",
       "22           Sapajus_apella   3    male      816          411    0.642157   \n",
       "23           Sapajus_apella   3  female     4100         1575    0.856829   \n",
       "24  Rhinopithecus_roxellana   4    male    54115        29342    0.417186   \n",
       "25  Rhinopithecus_roxellana   4  female   281576       114378    0.537187   \n",
       "26   Trachypithecus_phayrei   1    male    10339        10339    1.000000   \n",
       "27   Trachypithecus_phayrei   1  female    20475        20475    1.000000   \n",
       "\n",
       "    all_shared_het  above_hw  weight_above_hw  total_het  both_f_m   het_all  \\\n",
       "0         0.003908  0.080349         0.321127     399595  0.091703  0.000015   \n",
       "1         0.000050  0.008458         0.044298     399595  0.091703  0.000015   \n",
       "2         0.126200  0.284394         0.587242     105367  0.084761  0.003568   \n",
       "3         0.009317  0.103327         0.271532     105367  0.084761  0.003568   \n",
       "4         0.000889  0.016185         0.091210     686421  0.087024  0.000000   \n",
       "5         0.000002  0.001205         0.010409     686421  0.087024  0.000000   \n",
       "6         0.002359  0.065884         0.217316      60073  0.102375  0.000166   \n",
       "7         0.000799  0.004225         0.017854      60073  0.102375  0.000166   \n",
       "8         1.000000  0.000000         0.000000     143761  0.015790  0.015790   \n",
       "9         1.000000  0.000000         0.000000     143761  0.015790  0.015790   \n",
       "10        0.352794  0.496590         0.727194     551534  0.032045  0.000225   \n",
       "11        0.013280  0.086731         0.183705     551534  0.032045  0.000225   \n",
       "12        0.312562  0.312562         0.505179      27102  0.133496  0.019777   \n",
       "13        0.038726  0.038726         0.091329      27102  0.133496  0.019777   \n",
       "14        0.001539  0.054680         0.299281    1916367  0.041915  0.000002   \n",
       "15        0.000013  0.002071         0.021359    1916367  0.041915  0.000002   \n",
       "16        1.000000  0.000000         0.000000     180589  0.140989  0.140989   \n",
       "17        1.000000  0.000000         0.000000     180589  0.140989  0.140989   \n",
       "18        0.340236  0.340236         0.507725     250509  0.022893  0.003249   \n",
       "19        0.029005  0.029005         0.056376     250509  0.022893  0.003249   \n",
       "20        0.247255  0.247255         0.436321      21480  0.050047  0.002793   \n",
       "21        0.028011  0.028011         0.066358      21480  0.050047  0.002793   \n",
       "22        0.154412  0.154412         0.306321       4607  0.067072  0.003690   \n",
       "23        0.009268  0.009268         0.024127       4607  0.067072  0.003690   \n",
       "24        0.217555  0.368548         0.610074     319596  0.050360  0.000729   \n",
       "25        0.001868  0.160149         0.296840     319596  0.050360  0.000729   \n",
       "26        1.000000  0.000000         0.000000      26935  0.144013  0.144013   \n",
       "27        1.000000  0.000000         0.000000      26935  0.144013  0.144013   \n",
       "\n",
       "    above_hw_m_f  \n",
       "0       0.014802  \n",
       "1       0.014802  \n",
       "2       0.035172  \n",
       "3       0.035172  \n",
       "4       0.002785  \n",
       "5       0.002785  \n",
       "6       0.017063  \n",
       "7       0.017063  \n",
       "8       0.000000  \n",
       "9       0.000000  \n",
       "10      0.005882  \n",
       "11      0.005882  \n",
       "12      0.059848  \n",
       "13      0.059848  \n",
       "14      0.002658  \n",
       "15      0.002658  \n",
       "16      0.000000  \n",
       "17      0.000000  \n",
       "18      0.010071  \n",
       "19      0.010071  \n",
       "20      0.017691  \n",
       "21      0.017691  \n",
       "22      0.020404  \n",
       "23      0.020404  \n",
       "24      0.022488  \n",
       "25      0.022488  \n",
       "26      0.000000  \n",
       "27      0.000000  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat(df_list, ignore_index=True)\n",
    "df.reset_index()\n",
    "df[\"average_het\"] = df[\"average_het\"].astype(int)\n",
    "df[\"s_het\"] = df[\"s_het\"].astype(int)\n",
    "df[\"total_het\"] = df[\"total_het\"].astype(int)\n",
    "df.to_csv(\"../results/primate_diversity_het_statistics.txt\", header=True,\n",
    "          index=False, sep='\\t', mode='a', float_format='%.5f')\n",
    "df.to_csv(\"../results/primate_diversity_het_statistics.csv\", header=True,\n",
    "          index=False, float_format='%.5f')\n",
    "df.to_html(\"../results/primate_diversity_het_statistics.html\", float_format='%.5f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>both_f_m</th>\n",
       "      <th>het_all</th>\n",
       "      <th>above_hw_m_f</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.091703</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.014802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.091703</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.014802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.084761</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.035172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.084761</td>\n",
       "      <td>0.003568</td>\n",
       "      <td>0.035172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.087024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.087024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.002785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.102375</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.017063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.102375</td>\n",
       "      <td>0.000166</td>\n",
       "      <td>0.017063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.015790</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.032045</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0.032045</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.005882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0.133496</td>\n",
       "      <td>0.019777</td>\n",
       "      <td>0.059848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>0.133496</td>\n",
       "      <td>0.019777</td>\n",
       "      <td>0.059848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>0.041915</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.002658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.140989</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>0.022893</td>\n",
       "      <td>0.003249</td>\n",
       "      <td>0.010071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.022893</td>\n",
       "      <td>0.003249</td>\n",
       "      <td>0.010071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.050047</td>\n",
       "      <td>0.002793</td>\n",
       "      <td>0.017691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0.050047</td>\n",
       "      <td>0.002793</td>\n",
       "      <td>0.017691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0.067072</td>\n",
       "      <td>0.003690</td>\n",
       "      <td>0.020404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>0.067072</td>\n",
       "      <td>0.003690</td>\n",
       "      <td>0.020404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>0.050360</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.022488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>0.050360</td>\n",
       "      <td>0.000729</td>\n",
       "      <td>0.022488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    both_f_m   het_all  above_hw_m_f\n",
       "0   0.091703  0.000015      0.014802\n",
       "1   0.091703  0.000015      0.014802\n",
       "2   0.084761  0.003568      0.035172\n",
       "3   0.084761  0.003568      0.035172\n",
       "4   0.087024  0.000000      0.002785\n",
       "5   0.087024  0.000000      0.002785\n",
       "6   0.102375  0.000166      0.017063\n",
       "7   0.102375  0.000166      0.017063\n",
       "8   0.015790  0.015790      0.000000\n",
       "9   0.015790  0.015790      0.000000\n",
       "10  0.032045  0.000225      0.005882\n",
       "11  0.032045  0.000225      0.005882\n",
       "12  0.133496  0.019777      0.059848\n",
       "13  0.133496  0.019777      0.059848\n",
       "14  0.041915  0.000002      0.002658\n",
       "15  0.041915  0.000002      0.002658\n",
       "16  0.140989  0.140989      0.000000\n",
       "17  0.140989  0.140989      0.000000\n",
       "18  0.022893  0.003249      0.010071\n",
       "19  0.022893  0.003249      0.010071\n",
       "20  0.050047  0.002793      0.017691\n",
       "21  0.050047  0.002793      0.017691\n",
       "22  0.067072  0.003690      0.020404\n",
       "23  0.067072  0.003690      0.020404\n",
       "24  0.050360  0.000729      0.022488\n",
       "25  0.050360  0.000729      0.022488\n",
       "26  0.144013  0.144013      0.000000\n",
       "27  0.144013  0.144013      0.000000"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[:, 10:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6 4\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[     0., 144346.,  80228.,  40864.,     43.],\n",
       "       [ 17578.,   2977.,   1322.,    657.,     42.],\n",
       "       [  7685.,   1737.,   1285.,    809.,     79.],\n",
       "       [  5063.,   1090.,   1046.,    843.,    129.],\n",
       "       [  7694.,   1109.,   1342.,   1395.,    233.]])"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genus = \"Rhinopithecus_roxellana.fasta\"\n",
    "d = d_hetero.loc[(d_hetero.PDGP_ID != \"PD_0001\") & (d_hetero.PDGP_ID != \"PD_0368\")]\n",
    "d = d.loc[d.reference_file == genus]\n",
    "n_matrix = generate_shared_het_matrix(d, genus_dict[genus])\n",
    "n_matrix"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
